{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/cristianmejia00/clustering/blob/main/Topic_Models_using_BERTopic_TOPIC_MODEL_20241101.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Zj3lYckpO2Yn"
      },
      "source": [
        "# Topic Modeling with BERTopic\n",
        "\n",
        "🔴 copied from the [Kubota Colab](https://colab.research.google.com/drive/1YsDp5_qGXGJKsEXsS8DO8CA_lqZc6EpA).  \n",
        "\n",
        "`Topic Models` are methods to automatically organize a corpus of text into topics.\n",
        "\n",
        "Topic Model process:\n",
        "1. Data preparation\n",
        "2. Tranform text to numeric vectors\n",
        "3. Multidimensionality reduction\n",
        "4. Clustering\n",
        "5. Topic analysis\n",
        "6. Cluster assignation\n",
        "\n",
        "\n",
        "This notebook uses the library `BERTopic` which is a one-stop solution for topic modeling including handy functions for plotting and analysis. However, BERTopic does not have a function to extract the X and Y coords from UMAP. If we need the coordinates then use the notebooks `Topic_Models_using_Transformers` instead. In any other situation, when a quick analysis is needed this notebook may be better.\n",
        "\n",
        "This notebook is also the one needed for the heatmap codes included in this folder.\n",
        "\n",
        "`BERTopic` is Python library that handles steps 2 to 6.\n",
        "BERT topic models use the transformer architechture to generate the embeds (i.e. the vector or numeric representation of words) and are currently the state-of-the-art method for vectorization.\n",
        "\n",
        "This notebook shows how to use it.\n",
        "\n",
        "---\n",
        "Reading:\n",
        "[Topic Modeling with Deep Learning Using Python BERTopic](https://medium.com/grabngoinfo/topic-modeling-with-deep-learning-using-python-bertopic-cf91f5676504)\n",
        "[Advanced Topic Modeling with BERTopic](https://www.pinecone.io/learn/bertopic/)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WlufmOowbHSF"
      },
      "source": [
        "# Requirements"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kFQU-BC5ReYd"
      },
      "source": [
        "## Packages installation and initialization"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ckx8hOIyRZvv",
        "outputId": "20e7642f-824b-4ba1-86cc-5a97c89670c5",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting bertopic[visualization]\n",
            "  Downloading bertopic-0.16.4-py3-none-any.whl.metadata (23 kB)\n",
            "\u001b[33mWARNING: bertopic 0.16.4 does not provide the extra 'visualization'\u001b[0m\u001b[33m\n",
            "\u001b[0mCollecting hdbscan>=0.8.29 (from bertopic[visualization])\n",
            "  Downloading hdbscan-0.8.39-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (15 kB)\n",
            "Requirement already satisfied: numpy>=1.20.0 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (1.26.4)\n",
            "Requirement already satisfied: pandas>=1.1.5 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (2.2.2)\n",
            "Requirement already satisfied: plotly>=4.7.0 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (5.24.1)\n",
            "Requirement already satisfied: scikit-learn>=0.22.2.post1 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (1.5.2)\n",
            "Requirement already satisfied: sentence-transformers>=0.4.1 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (3.2.1)\n",
            "Requirement already satisfied: tqdm>=4.41.1 in /usr/local/lib/python3.10/dist-packages (from bertopic[visualization]) (4.66.6)\n",
            "Collecting umap-learn>=0.5.0 (from bertopic[visualization])\n",
            "  Downloading umap_learn-0.5.7-py3-none-any.whl.metadata (21 kB)\n",
            "Requirement already satisfied: scipy>=1.0 in /usr/local/lib/python3.10/dist-packages (from hdbscan>=0.8.29->bertopic[visualization]) (1.13.1)\n",
            "Requirement already satisfied: joblib>=1.0 in /usr/local/lib/python3.10/dist-packages (from hdbscan>=0.8.29->bertopic[visualization]) (1.4.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.1.5->bertopic[visualization]) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.1.5->bertopic[visualization]) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas>=1.1.5->bertopic[visualization]) (2024.2)\n",
            "Requirement already satisfied: tenacity>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from plotly>=4.7.0->bertopic[visualization]) (9.0.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from plotly>=4.7.0->bertopic[visualization]) (24.1)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.22.2.post1->bertopic[visualization]) (3.5.0)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.41.0 in /usr/local/lib/python3.10/dist-packages (from sentence-transformers>=0.4.1->bertopic[visualization]) (4.44.2)\n",
            "Requirement already satisfied: torch>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from sentence-transformers>=0.4.1->bertopic[visualization]) (2.5.0+cu121)\n",
            "Requirement already satisfied: huggingface-hub>=0.20.0 in /usr/local/lib/python3.10/dist-packages (from sentence-transformers>=0.4.1->bertopic[visualization]) (0.24.7)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.10/dist-packages (from sentence-transformers>=0.4.1->bertopic[visualization]) (10.4.0)\n",
            "Requirement already satisfied: numba>=0.51.2 in /usr/local/lib/python3.10/dist-packages (from umap-learn>=0.5.0->bertopic[visualization]) (0.60.0)\n",
            "Collecting pynndescent>=0.5 (from umap-learn>=0.5.0->bertopic[visualization])\n",
            "  Downloading pynndescent-0.5.13-py3-none-any.whl.metadata (6.8 kB)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.16.1)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (2024.10.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (6.0.2)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (2.32.3)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (4.12.2)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.51.2->umap-learn>=0.5.0->bertopic[visualization]) (0.43.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas>=1.1.5->bertopic[visualization]) (1.16.0)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.1.4)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.10/dist-packages (from torch>=1.11.0->sentence-transformers>=0.4.1->bertopic[visualization]) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from sympy==1.13.1->torch>=1.11.0->sentence-transformers>=0.4.1->bertopic[visualization]) (1.3.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic[visualization]) (2024.9.11)\n",
            "Requirement already satisfied: safetensors>=0.4.1 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic[visualization]) (0.4.5)\n",
            "Requirement already satisfied: tokenizers<0.20,>=0.19 in /usr/local/lib/python3.10/dist-packages (from transformers<5.0.0,>=4.41.0->sentence-transformers>=0.4.1->bertopic[visualization]) (0.19.1)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch>=1.11.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.0.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.20.0->sentence-transformers>=0.4.1->bertopic[visualization]) (2024.8.30)\n",
            "Downloading hdbscan-0.8.39-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.2/4.2 MB\u001b[0m \u001b[31m21.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading umap_learn-0.5.7-py3-none-any.whl (88 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m88.8/88.8 kB\u001b[0m \u001b[31m3.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading bertopic-0.16.4-py3-none-any.whl (143 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m143.7/143.7 kB\u001b[0m \u001b[31m3.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pynndescent-0.5.13-py3-none-any.whl (56 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.9/56.9 kB\u001b[0m \u001b[31m3.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: pynndescent, hdbscan, umap-learn, bertopic\n",
            "Successfully installed bertopic-0.16.4 hdbscan-0.8.39 pynndescent-0.5.13 umap-learn-0.5.7\n"
          ]
        }
      ],
      "source": [
        "!pip install bertopic[visualization]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "gMlZ_7DkOxGG"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import time\n",
        "import math\n",
        "import uuid\n",
        "import re\n",
        "import os\n",
        "import json\n",
        "import pickle\n",
        "from datetime import date\n",
        "from itertools import compress\n",
        "from bertopic import BERTopic\n",
        "from umap import UMAP\n",
        "from gensim.parsing.preprocessing import remove_stopwords\n",
        "from sklearn.cluster import KMeans\n",
        "from sentence_transformers import SentenceTransformer"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SZ1BccaqRtS0"
      },
      "source": [
        "## Connect your Google Drive"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6os5BGPFRxfX",
        "outputId": "41e02c25-1dc9-426e-f441-874fc600fc70"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def find_e_keys(dictionary):\n",
        "    # List comprehension to find keys starting with 'e'\n",
        "    e_keys = [key for key in dictionary if str(key).lower().startswith('e')]\n",
        "    return e_keys"
      ],
      "metadata": {
        "id": "f2p9sY7sUYkU"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dDm5-r5oR98q"
      },
      "source": [
        "# 🔴 Input files and options\n",
        "\n",
        "Go to your Google Drive and create a folder in the root directory. We are going to save all related data in that directory.\n",
        "Upload the dataset of news into the above folder.\n",
        "- The dataset should be a `.csv` file.\n",
        "- Every row in the dataset is a document\n",
        "- It can any kind of columns. Some columns must contain the text we want to analyze. For example, a dataset of academic articles may contain a \"Title\" and/or \"Abstract\" column."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "5ICq5zQbSJa8"
      },
      "outputs": [],
      "source": [
        "# The bibliometrics folder\n",
        "ROOT_FOLDER_PATH = f\"drive/MyDrive/Bibliometrics_Drive\"\n",
        "\n",
        "# Change to the name of the folder where the dataset is uploaded inside the above folder\n",
        "project_folder = 'Q312_utokyo'\n",
        "\n",
        "analysis_id = 'a01_tm__f01_e01__km01'\n",
        "\n",
        "# Filtered label\n",
        "settings_directive = \"settings_analysis_directive_2024-11-02-15-36.json\""
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Read settings\n",
        "with open(f'{ROOT_FOLDER_PATH}/{project_folder}/{analysis_id}/{settings_directive}', 'r') as file:\n",
        "    settings = json.load(file)"
      ],
      "metadata": {
        "id": "QpoR8P3O0ACf"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Input dataset\n",
        "dataset_file_path = f\"{ROOT_FOLDER_PATH}/{settings['metadata']['project_folder']}/{settings['metadata']['filtered_folder']}/dataset_raw_cleaned.csv\""
      ],
      "metadata": {
        "id": "COh3zc3b2oAn"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "biEFdxrm0WrO"
      },
      "outputs": [],
      "source": [
        "# Function to save files\n",
        "def save_as_csv(df, save_name_without_extension, with_index):\n",
        "    \"usage: `save_as_csv(dataframe, 'filename')`\"\n",
        "    df.to_csv(f\"{ROOT_FOLDER_PATH}/{save_name_without_extension}.csv\", index=with_index)\n",
        "    print(\"===\\nSaved: \", f\"{ROOT_FOLDER_PATH}/{save_name_without_extension}.csv\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "Ff7hQXE1dYDw"
      },
      "outputs": [],
      "source": [
        "# prompt: a function to save object to a pickle file\n",
        "def save_object_as_pickle(obj, filename):\n",
        "  \"\"\"\n",
        "  Saves an object as a pickle file.\n",
        "\n",
        "  Args:\n",
        "      obj: The object to be saved.\n",
        "      filename: The filename of the pickle file.\n",
        "  \"\"\"\n",
        "  with open(filename, \"wb\") as f:\n",
        "    pickle.dump(obj, f)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "iljQe0_xb2FU"
      },
      "outputs": [],
      "source": [
        "# prompt: a function to load pickle object given a path\n",
        "def load_pickle(path):\n",
        "    with open(path, 'rb') as f:\n",
        "        return pickle.load(f)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 655
        },
        "id": "Tob-4BIaUbZ9",
        "outputId": "d48eac6d-2af8-4991-af19-5bda2386e6bc"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(11120, 39)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   X_N                                  uuid PT  \\\n",
              "0    1  41e14258-945f-4ec9-9e0a-e58c87752c10  J   \n",
              "1    2  f0daa586-2560-43ef-8132-f3b87f9167a3  J   \n",
              "2    3  23327f73-de46-43e9-8796-decc6950c06d  J   \n",
              "3    4  dff88211-4a40-4566-a7c0-1856b531092f  J   \n",
              "4    5  5e23ca4b-52e5-47ad-9812-1e7a65f2b2a8  J   \n",
              "\n",
              "                                                  AU  \\\n",
              "0  Kishi, T; Kobayashi, K; Sasagawa, K; Sakimura,...   \n",
              "1                              Berenger, F; Tsuda, K   \n",
              "2                                    Kato, S; Ono, S   \n",
              "3  Kojima, K; Chambers, JK; Yoshizawa, M; Fujioka...   \n",
              "4          Delgado-Munoz, J; Matsunaka, R; Hiraki, K   \n",
              "\n",
              "                                                  TI  \\\n",
              "0  Automated analysis of a novel object recogniti...   \n",
              "1  An ANI-2 enabled open-source protocol to estim...   \n",
              "2  Accelerated Approval for Cancer Drugs in the U...   \n",
              "3  Pathological features of intrathoracic histioc...   \n",
              "4  Classification of Known and Unknown Study Item...   \n",
              "\n",
              "                                      SO       LA                     DT  \\\n",
              "0             BEHAVIOURAL BRAIN RESEARCH  English                Article   \n",
              "1     JOURNAL OF COMPUTATIONAL CHEMISTRY  English  Article; Early Access   \n",
              "2   JOURNAL OF PHARMACEUTICAL INNOVATION  English                Article   \n",
              "3  JOURNAL OF VETERINARY MEDICAL SCIENCE  English                Article   \n",
              "4                         BRAIN SCIENCES  English                Article   \n",
              "\n",
              "                                                  DE  \\\n",
              "0  Animal behavior; Novel object recognition test...   \n",
              "1  ANI-2x; docking; ligand; LIT-PCBA; MC; QM; SBV...   \n",
              "2  Accelerated approval; Oncology; Evidence and a...   \n",
              "3  histiocytic sarcoma; spiny rat; Tokudaia osime...   \n",
              "4  long term memory; familiarity; electroencephal...   \n",
              "\n",
              "                                            ID  ...   VL   IS    BP    EP  \\\n",
              "0                              ANXIETY; MODELS  ...  476  NaN   NaN   NaN   \n",
              "1             OPEN-SOURCE PACKAGE; FORCE-FIELD  ...  NaN  NaN   NaN   NaN   \n",
              "2                         END-POINTS; ONCOLOGY  ...   19    5   NaN   NaN   \n",
              "3                                   CHROMOSOME  ...   86   10  1076  1080   \n",
              "4  ERP; RECOGNITION; FAMILIARITY; RECOLLECTION  ...   14    9   NaN   NaN   \n",
              "\n",
              "       AR                          DI  PG                                  WC  \\\n",
              "0  115278   10.1016/j.bbr.2024.115278   5  Behavioral Sciences; Neurosciences   \n",
              "1     NaN           10.1002/jcc.27478  11        Chemistry, Multidisciplinary   \n",
              "2      54  10.1007/s12247-024-09851-9  12             Pharmacology & Pharmacy   \n",
              "3     NaN        10.1292/jvms.24-0185   5                 Veterinary Sciences   \n",
              "4     860    10.3390/brainsci14090860  21                       Neurosciences   \n",
              "\n",
              "                                               SC                   UT  \n",
              "0  Behavioral Sciences; Neurosciences & Neurology  WOS:001333148000001  \n",
              "1                                       Chemistry  WOS:001328968200001  \n",
              "2                         Pharmacology & Pharmacy  WOS:001304011000001  \n",
              "3                             Veterinary Sciences  WOS:001338099800007  \n",
              "4                       Neurosciences & Neurology  WOS:001323906200001  \n",
              "\n",
              "[5 rows x 39 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ff8ff84e-9d37-4eec-afb9-659ceb9eebbb\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>X_N</th>\n",
              "      <th>uuid</th>\n",
              "      <th>PT</th>\n",
              "      <th>AU</th>\n",
              "      <th>TI</th>\n",
              "      <th>SO</th>\n",
              "      <th>LA</th>\n",
              "      <th>DT</th>\n",
              "      <th>DE</th>\n",
              "      <th>ID</th>\n",
              "      <th>...</th>\n",
              "      <th>VL</th>\n",
              "      <th>IS</th>\n",
              "      <th>BP</th>\n",
              "      <th>EP</th>\n",
              "      <th>AR</th>\n",
              "      <th>DI</th>\n",
              "      <th>PG</th>\n",
              "      <th>WC</th>\n",
              "      <th>SC</th>\n",
              "      <th>UT</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>41e14258-945f-4ec9-9e0a-e58c87752c10</td>\n",
              "      <td>J</td>\n",
              "      <td>Kishi, T; Kobayashi, K; Sasagawa, K; Sakimura,...</td>\n",
              "      <td>Automated analysis of a novel object recogniti...</td>\n",
              "      <td>BEHAVIOURAL BRAIN RESEARCH</td>\n",
              "      <td>English</td>\n",
              "      <td>Article</td>\n",
              "      <td>Animal behavior; Novel object recognition test...</td>\n",
              "      <td>ANXIETY; MODELS</td>\n",
              "      <td>...</td>\n",
              "      <td>476</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>115278</td>\n",
              "      <td>10.1016/j.bbr.2024.115278</td>\n",
              "      <td>5</td>\n",
              "      <td>Behavioral Sciences; Neurosciences</td>\n",
              "      <td>Behavioral Sciences; Neurosciences &amp; Neurology</td>\n",
              "      <td>WOS:001333148000001</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>2</td>\n",
              "      <td>f0daa586-2560-43ef-8132-f3b87f9167a3</td>\n",
              "      <td>J</td>\n",
              "      <td>Berenger, F; Tsuda, K</td>\n",
              "      <td>An ANI-2 enabled open-source protocol to estim...</td>\n",
              "      <td>JOURNAL OF COMPUTATIONAL CHEMISTRY</td>\n",
              "      <td>English</td>\n",
              "      <td>Article; Early Access</td>\n",
              "      <td>ANI-2x; docking; ligand; LIT-PCBA; MC; QM; SBV...</td>\n",
              "      <td>OPEN-SOURCE PACKAGE; FORCE-FIELD</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>10.1002/jcc.27478</td>\n",
              "      <td>11</td>\n",
              "      <td>Chemistry, Multidisciplinary</td>\n",
              "      <td>Chemistry</td>\n",
              "      <td>WOS:001328968200001</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>3</td>\n",
              "      <td>23327f73-de46-43e9-8796-decc6950c06d</td>\n",
              "      <td>J</td>\n",
              "      <td>Kato, S; Ono, S</td>\n",
              "      <td>Accelerated Approval for Cancer Drugs in the U...</td>\n",
              "      <td>JOURNAL OF PHARMACEUTICAL INNOVATION</td>\n",
              "      <td>English</td>\n",
              "      <td>Article</td>\n",
              "      <td>Accelerated approval; Oncology; Evidence and a...</td>\n",
              "      <td>END-POINTS; ONCOLOGY</td>\n",
              "      <td>...</td>\n",
              "      <td>19</td>\n",
              "      <td>5</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>54</td>\n",
              "      <td>10.1007/s12247-024-09851-9</td>\n",
              "      <td>12</td>\n",
              "      <td>Pharmacology &amp; Pharmacy</td>\n",
              "      <td>Pharmacology &amp; Pharmacy</td>\n",
              "      <td>WOS:001304011000001</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>4</td>\n",
              "      <td>dff88211-4a40-4566-a7c0-1856b531092f</td>\n",
              "      <td>J</td>\n",
              "      <td>Kojima, K; Chambers, JK; Yoshizawa, M; Fujioka...</td>\n",
              "      <td>Pathological features of intrathoracic histioc...</td>\n",
              "      <td>JOURNAL OF VETERINARY MEDICAL SCIENCE</td>\n",
              "      <td>English</td>\n",
              "      <td>Article</td>\n",
              "      <td>histiocytic sarcoma; spiny rat; Tokudaia osime...</td>\n",
              "      <td>CHROMOSOME</td>\n",
              "      <td>...</td>\n",
              "      <td>86</td>\n",
              "      <td>10</td>\n",
              "      <td>1076</td>\n",
              "      <td>1080</td>\n",
              "      <td>NaN</td>\n",
              "      <td>10.1292/jvms.24-0185</td>\n",
              "      <td>5</td>\n",
              "      <td>Veterinary Sciences</td>\n",
              "      <td>Veterinary Sciences</td>\n",
              "      <td>WOS:001338099800007</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>5</td>\n",
              "      <td>5e23ca4b-52e5-47ad-9812-1e7a65f2b2a8</td>\n",
              "      <td>J</td>\n",
              "      <td>Delgado-Munoz, J; Matsunaka, R; Hiraki, K</td>\n",
              "      <td>Classification of Known and Unknown Study Item...</td>\n",
              "      <td>BRAIN SCIENCES</td>\n",
              "      <td>English</td>\n",
              "      <td>Article</td>\n",
              "      <td>long term memory; familiarity; electroencephal...</td>\n",
              "      <td>ERP; RECOGNITION; FAMILIARITY; RECOLLECTION</td>\n",
              "      <td>...</td>\n",
              "      <td>14</td>\n",
              "      <td>9</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>860</td>\n",
              "      <td>10.3390/brainsci14090860</td>\n",
              "      <td>21</td>\n",
              "      <td>Neurosciences</td>\n",
              "      <td>Neurosciences &amp; Neurology</td>\n",
              "      <td>WOS:001323906200001</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 39 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ff8ff84e-9d37-4eec-afb9-659ceb9eebbb')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ff8ff84e-9d37-4eec-afb9-659ceb9eebbb button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ff8ff84e-9d37-4eec-afb9-659ceb9eebbb');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-6bb2201d-8d79-4228-bec3-947704831495\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-6bb2201d-8d79-4228-bec3-947704831495')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-6bb2201d-8d79-4228-bec3-947704831495 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 13
        }
      ],
      "source": [
        "# Open the data file\n",
        "df = pd.read_csv(f\"{dataset_file_path}\")\n",
        "print(df.shape)\n",
        "df.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9UrPiLepN1s9"
      },
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## PART 2: Topic Model"
      ],
      "metadata": {
        "id": "eK-U-Sl8fc-m"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "FhwGnBjFQT4x"
      },
      "outputs": [],
      "source": [
        "# bibliometrics_folder\n",
        "# project_folder\n",
        "# project_name_suffix\n",
        "# ROOT_FOLDER_PATH = f\"drive/MyDrive/{bibliometrics_folder}\"\n",
        "\n",
        "#############################################################\n",
        "# Embeddings folder\n",
        "embeddings_folder_name = settings['tmo']['embeds_folder']\n",
        "\n",
        "# Which column has the year of the documents?\n",
        "my_year = settings['tmo']['year_column']\n",
        "\n",
        "# Number of topics. Select the number of topics to extract.\n",
        "# Choose 0, for automatic detection.\n",
        "n_topics = settings['tmo']['n_topics']\n",
        "\n",
        "# Minimum number of documents per topic\n",
        "min_topic_size = settings['tmo']['min_topic_size']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "9RYSioo2cxxC"
      },
      "outputs": [],
      "source": [
        "# Get the embeddings back.\n",
        "embeddings = load_pickle(f\"{ROOT_FOLDER_PATH}/{project_folder}/{settings['metadata']['filtered_folder']}/{embeddings_folder_name}/embeddings.pck\")\n",
        "corpus =     pd.read_csv(f\"{ROOT_FOLDER_PATH}/{project_folder}/{settings['metadata']['filtered_folder']}/{embeddings_folder_name}/corpus.csv\").reset_index(drop=True)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "injCV_IPN26L"
      },
      "outputs": [],
      "source": [
        "# Combine embeddings\n",
        "documents = corpus.text.to_list()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# corpus['uuid'] = [uuid.uuid4() for _ in range(len(corpus.index))]\n",
        "# corpus['X_N'] = [i for i in range(1, len(corpus.index)+1)]"
      ],
      "metadata": {
        "id": "ve_oloYNrBLz"
      },
      "execution_count": 38,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "len(documents)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "onVFOjBZuMmJ",
        "outputId": "4898b936-f4ea-4a7d-8448-a65466de85f2"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "11120"
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "S7LeuG_7S4Db",
        "outputId": "4e27c106-2e0c-4630-a741-2fbd5e5512b0"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ],
      "source": [
        "#len(embeddings) == len(documents)\n",
        "len(embeddings['embeddings']) == len(documents)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "n6qUvhGGcPZo"
      },
      "outputs": [],
      "source": [
        "from hdbscan.hdbscan_ import HDBSCAN\n",
        "# Execute the topic model.\n",
        "# I suggest changing the values marked with #<---\n",
        "# The others are the default values and they'll work fine in most cases.\n",
        "# This will take several minutes to finish.\n",
        "\n",
        "# Initiate UMAP\n",
        "umap_model = UMAP(n_neighbors=15,\n",
        "                  n_components=5,\n",
        "                  min_dist=0.0,\n",
        "                  metric='cosine',\n",
        "                  random_state=100)\n",
        "\n",
        "if n_topics == 0:\n",
        "  # Initiate topic model with HDBScan (Automatic topic selection)\n",
        "  topic_model_params = HDBSCAN(min_cluster_size=min_topic_size,\n",
        "                               metric='euclidean',\n",
        "                               cluster_selection_method='eom',\n",
        "                               prediction_data=True)\n",
        "else:\n",
        "  # Initiate topic model with K-means (Manual topic selection)\n",
        "  topic_model_params = KMeans(n_clusters = n_topics)\n",
        "\n",
        "# Initiate BERTopic\n",
        "topic_model = BERTopic(umap_model = umap_model,\n",
        "                       hdbscan_model = topic_model_params,\n",
        "                       min_topic_size=min_topic_size,\n",
        "                       #nr_topics=15,          #<--- Footnote 1\n",
        "                       n_gram_range=(1,3),\n",
        "                       language='english',\n",
        "                       calculate_probabilities=True,\n",
        "                       verbose=True)\n",
        "\n",
        "\n",
        "\n",
        "# Footnote 1: This controls the number of topics we want AFTER clustering.\n",
        "# Add a hashtag at the beggining to use the number of topics returned by the topic model.\n",
        "# When using HDBScan nr_topics will be obtained after orphans removal, and there is no warranty that `nr_topics < HDBScan topics`.\n",
        "# thus, with HDBScan `nr_topics` means N topics OR LESS.\n",
        "# When using KMeans nr_topics can be used to further reduce the number of topics.\n",
        "# We use the topics as returned by the topic model. So we do not need to activate it here."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {
        "id": "FQQsmPGzUCuB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a2b43bac-051a-4599-ad04-a36b1a575a26"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "2024-11-02 06:41:31,593 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm\n",
            "2024-11-02 06:42:16,351 - BERTopic - Dimensionality - Completed ✓\n",
            "2024-11-02 06:42:16,354 - BERTopic - Cluster - Start clustering the reduced embeddings\n",
            "2024-11-02 06:42:17,078 - BERTopic - Cluster - Completed ✓\n",
            "2024-11-02 06:42:17,098 - BERTopic - Representation - Extracting topics from clusters using representation models.\n",
            "2024-11-02 06:42:45,870 - BERTopic - Representation - Completed ✓\n"
          ]
        }
      ],
      "source": [
        "# Compute topic model\n",
        "#topics, probabilities = topic_model.fit_transform(documents, embeddings)\n",
        "topics, probabilities = topic_model.fit_transform(documents, embeddings['embeddings'])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "67pgsMtwsFmH",
        "outputId": "82c86e95-6be7-4832-a076-94d8d9dcd8ae"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Topic  Count                                               Name  \\\n",
              "0        0    279                         0_magnetic_spin_field_hall   \n",
              "1        1    261                     1_galaxies_similar_galaxy_mass   \n",
              "2        2    229                          2_cells_cancer_cell_tumor   \n",
              "3        3    214  3_theory_usepackagemathrsfs usepackageupgreek ...   \n",
              "4        4    200                          4_patients_95_outcomes_ci   \n",
              "..     ...    ...                                                ...   \n",
              "107    107     38            107_patients_fgf23_hearing_hearing loss   \n",
              "108    108     35              108_psii_rhodopsins_chromophore_light   \n",
              "109    109     35           109_equations_problem_solutions_solution   \n",
              "110    110     34            110_asteroid_phobos_asteroids_dimorphos   \n",
              "111    111     31                     111_kuroshio_coastal_delta_14c   \n",
              "\n",
              "                                        Representation  \\\n",
              "0    [magnetic, spin, field, hall, magnetization, p...   \n",
              "1    [galaxies, similar, galaxy, mass, stellar, sta...   \n",
              "2    [cells, cancer, cell, tumor, expression, immun...   \n",
              "3    [theory, usepackagemathrsfs usepackageupgreek ...   \n",
              "4    [patients, 95, outcomes, ci, risk, heart, arte...   \n",
              "..                                                 ...   \n",
              "107  [patients, fgf23, hearing, hearing loss, pain,...   \n",
              "108  [psii, rhodopsins, chromophore, light, rhodops...   \n",
              "109  [equations, problem, solutions, solution, equa...   \n",
              "110  [asteroid, phobos, asteroids, dimorphos, bould...   \n",
              "111  [kuroshio, coastal, delta, 14c, water, groundw...   \n",
              "\n",
              "                                   Representative_Docs  \n",
              "0    [magnetic anisotropy cecosi high magnetic fiel...  \n",
              "1    [photon burst clears earliest dusty galaxies: ...  \n",
              "2    [characterization kif20b novel prognostic biom...  \n",
              "3    [effective brane field theory higher-form symm...  \n",
              "4    [clinical outcomes mitral valve surgery atrial...  \n",
              "..                                                 ...  \n",
              "107  [case report: guanfacine methylphenidate impro...  \n",
              "108  [chromophore-protein interactions affecting po...  \n",
              "109  [characterization f-concavity preserved dirich...  \n",
              "110  [ultraviolet visible reflectance spectra phobo...  \n",
              "111  [impact large shallow semi-enclosed lagoon fre...  \n",
              "\n",
              "[112 rows x 5 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5e863f79-fc06-4ebb-80d0-0f9e43ee4185\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Topic</th>\n",
              "      <th>Count</th>\n",
              "      <th>Name</th>\n",
              "      <th>Representation</th>\n",
              "      <th>Representative_Docs</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>279</td>\n",
              "      <td>0_magnetic_spin_field_hall</td>\n",
              "      <td>[magnetic, spin, field, hall, magnetization, p...</td>\n",
              "      <td>[magnetic anisotropy cecosi high magnetic fiel...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>261</td>\n",
              "      <td>1_galaxies_similar_galaxy_mass</td>\n",
              "      <td>[galaxies, similar, galaxy, mass, stellar, sta...</td>\n",
              "      <td>[photon burst clears earliest dusty galaxies: ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>229</td>\n",
              "      <td>2_cells_cancer_cell_tumor</td>\n",
              "      <td>[cells, cancer, cell, tumor, expression, immun...</td>\n",
              "      <td>[characterization kif20b novel prognostic biom...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>214</td>\n",
              "      <td>3_theory_usepackagemathrsfs usepackageupgreek ...</td>\n",
              "      <td>[theory, usepackagemathrsfs usepackageupgreek ...</td>\n",
              "      <td>[effective brane field theory higher-form symm...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>200</td>\n",
              "      <td>4_patients_95_outcomes_ci</td>\n",
              "      <td>[patients, 95, outcomes, ci, risk, heart, arte...</td>\n",
              "      <td>[clinical outcomes mitral valve surgery atrial...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>107</th>\n",
              "      <td>107</td>\n",
              "      <td>38</td>\n",
              "      <td>107_patients_fgf23_hearing_hearing loss</td>\n",
              "      <td>[patients, fgf23, hearing, hearing loss, pain,...</td>\n",
              "      <td>[case report: guanfacine methylphenidate impro...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>108</th>\n",
              "      <td>108</td>\n",
              "      <td>35</td>\n",
              "      <td>108_psii_rhodopsins_chromophore_light</td>\n",
              "      <td>[psii, rhodopsins, chromophore, light, rhodops...</td>\n",
              "      <td>[chromophore-protein interactions affecting po...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>109</th>\n",
              "      <td>109</td>\n",
              "      <td>35</td>\n",
              "      <td>109_equations_problem_solutions_solution</td>\n",
              "      <td>[equations, problem, solutions, solution, equa...</td>\n",
              "      <td>[characterization f-concavity preserved dirich...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>110</th>\n",
              "      <td>110</td>\n",
              "      <td>34</td>\n",
              "      <td>110_asteroid_phobos_asteroids_dimorphos</td>\n",
              "      <td>[asteroid, phobos, asteroids, dimorphos, bould...</td>\n",
              "      <td>[ultraviolet visible reflectance spectra phobo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>111</th>\n",
              "      <td>111</td>\n",
              "      <td>31</td>\n",
              "      <td>111_kuroshio_coastal_delta_14c</td>\n",
              "      <td>[kuroshio, coastal, delta, 14c, water, groundw...</td>\n",
              "      <td>[impact large shallow semi-enclosed lagoon fre...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>112 rows × 5 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5e863f79-fc06-4ebb-80d0-0f9e43ee4185')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5e863f79-fc06-4ebb-80d0-0f9e43ee4185 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5e863f79-fc06-4ebb-80d0-0f9e43ee4185');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-48dae41a-dc5e-4209-aa7f-8097ff146395\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-48dae41a-dc5e-4209-aa7f-8097ff146395')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-48dae41a-dc5e-4209-aa7f-8097ff146395 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_54f14cb3-32b1-4d57-a00d-607c481d4757\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('tm_summary')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_54f14cb3-32b1-4d57-a00d-607c481d4757 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('tm_summary');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "tm_summary",
              "summary": "{\n  \"name\": \"tm_summary\",\n  \"rows\": 112,\n  \"fields\": [\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 32,\n        \"min\": 0,\n        \"max\": 111,\n        \"num_unique_values\": 112,\n        \"samples\": [\n          40,\n          65,\n          4\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Count\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 48,\n        \"min\": 31,\n        \"max\": 279,\n        \"num_unique_values\": 84,\n        \"samples\": [\n          50,\n          279,\n          68\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 112,\n        \"samples\": [\n          \"40_alloys_deformation_cutting_strain\",\n          \"65_flow_particle_particles_debris\",\n          \"4_patients_95_outcomes_ci\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 22
        }
      ],
      "source": [
        "# Get the list of topics\n",
        "# Topic = the topic number. From the largest topic.\n",
        "#         \"-1\" is the generic topic. Genericr keywords are aggegrated here.\n",
        "# Count = Documents assigned to this topic\n",
        "# Name = Top 4 words of the topic based on probability\n",
        "# Representation = The list of words representing this topic\n",
        "# Representative_Docs = Documents assigned to this topic\n",
        "tm_summary = topic_model.get_topic_info()\n",
        "tm_summary"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "id": "k9mN1tgZQupy"
      },
      "outputs": [],
      "source": [
        "# Save the topic model assets\n",
        "tm_folder_path = f'{ROOT_FOLDER_PATH}/{project_folder}/{settings[\"metadata\"][\"analysis_id\"]}'\n",
        "\n",
        "if not os.path.exists(tm_folder_path):\n",
        "  !mkdir $tm_folder_path\n",
        "\n",
        "tm_summary.to_csv(f'{tm_folder_path}/topic_model_info.csv', index=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 24,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WuE0vkbp3fVL",
        "outputId": "aa8ebf45-cdd3-4161-9ac4-1343c91f47da"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "112"
            ]
          },
          "metadata": {},
          "execution_count": 24
        }
      ],
      "source": [
        "# Number of topics found\n",
        "found_topics = max(tm_summary.Topic) + 1\n",
        "found_topics"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KjKUcNMdwFHQ",
        "outputId": "b3549b7c-4c8a-4b82-8d87-6a46774e4694"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 25
        }
      ],
      "source": [
        "# Confirm all documents are assigned\n",
        "sum(tm_summary.Count) == len(corpus)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bGyoQT47xY_R",
        "outputId": "9affafe5-0ff7-4edf-d9f6-3d7a7eda04eb"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('magnetic', 0.026596785516081942),\n",
              " ('spin', 0.01774999412542823),\n",
              " ('field', 0.008884241322670289),\n",
              " ('hall', 0.008818804380757694),\n",
              " ('magnetization', 0.008690627632162506),\n",
              " ('phase', 0.007211351594195054),\n",
              " ('effect', 0.007109006204478887),\n",
              " ('topological', 0.006577423864781797),\n",
              " ('magnetic field', 0.0063011913893084295),\n",
              " ('state', 0.006066045300763887)]"
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ],
      "source": [
        "# Get top 10 terms for a topic\n",
        "topic_model.get_topic(0)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PU99D2xhxoOY",
        "outputId": "4e90e928-9bbf-40c3-c3b5-a1a50abe82f0"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['magnetic anisotropy cecosi high magnetic field investigated high magnetic field properties ce-based kondo lattice cecosi means magnetization, magnetocaloric effect, contactless resistivity, specific heat, magnetic torque measurements pulsed high magnetic fields 55 t. cecosi proposed unconventional ordered phase called phase ii, appears paramagnetic phase antiferromagnetic phase iii. magnetization, magnetocaloric effect, contactless resistivity measurements determine field-angle dependence phase ii-iii boundary. magnetic torque measurements detect fourfold symmetry magnetic anisotropy high magnetic fields phases, implying magnetic hard axes (100) (110) plane tetragonal structure. comparison mean-field calculation magnetization data, determine crystal electric field energy scheme, agrees previous inelastic neutron scattering experiment. successfully reproduce fourfold magnetic anisotropy calculation, attributed hybridization crystal electric field wave functions. hybridization increases multipole degrees freedom origin unconventional order phase ii. moreover, specific heat 4 k exhibits nonphononic t3 term phase ii, indicates phase ii nota simple polarized paramagnetic state quadrupolar ordered state.',\n",
              " 'incommensurate magnetic order axion insulator candidate euin<sub>2</sub>as<sub>2</sub> investigated nmr measurement magnetic topological insulators exhibit unique electronic states interplay electronic topology spin structure. antiferromagnetic metal euin2as2 prominent candidate material exotic topological phases, including axion insulating state, theoretically predicted depending magnetic structure eu2+ moments. here, report experimental results nuclear magnetic resonance (nmr) measurements nuclei euin2as2 investigate coupling magnetic moments eu ions conduction electrons in2as2 layers magnetic structure. as-75 in-115 nmr spectra observed zero external magnetic fields reveal appearance internal fields 4.9 3.6 t, respectively, lowest temperature, suggesting strong coupling conduction electrons in2as2 layer ordered magnetic moments eu ions. as-75 nmr spectra in-plane external magnetic fields broad distributions internal fields produced incommensurate fan-like spin structure turns forced ferromagnetic state 0.7 t. propose spin reorientation process incommensurate helical state zero external magnetic field quickly changes fan state applying slight magnetic field.',\n",
              " \"electric-field control magnetic anisotropies: applications kitaev spin liquids topological spin textures magnetic anisotropies originate spin -orbit coupling determine magnetic ordering patterns. develop microscopic theory dc electric-field controls magnetic anisotropies magnetic mott insulators discuss applications kitaev materials topological spin textures. paper, microscopic approach based hubbard-type lattice models, tight-binding models onsite interactions. derive low-energy spin hamiltonian fourth-order perturbation expansion hubbard-type model. presence strong intra-atomic spin-orbit coupling dc electric fields add non-kitaev interactions dzyaloshinskii-moriya interaction off-diagonal gamma' interaction kitaev-heisenberg model induce topological quantum phase transition majorana chern insulating phases. investigate interatomic rashba spin-orbit coupling effects topological spin textures. dc electric fields turn create annihilate magnetic skyrmions, hedgehogs, chiral solitons. propose methods creating topological spin textures external electromagnetic fields. theory clarifies strong feasible electric field control kitaev spin liquids topological spin textures.\"]"
            ]
          },
          "metadata": {},
          "execution_count": 27
        }
      ],
      "source": [
        "# Get the top 10 documents for a topic\n",
        "topic_model.get_representative_docs(0)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1CgYkbHN9I2T"
      },
      "outputs": [],
      "source": [
        "# Others\n",
        "\n",
        "# # Get the number of documents per topic (same as in the table above)\n",
        "# topic_model.get_topic_freq(0)\n",
        "\n",
        "# # Get the main keywords per topic\n",
        "# topic_model.get_topics()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "JQKxUqHN9RT7",
        "outputId": "c572d73b-fc91-4c84-dc79-8f047b046aa2"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'calculate_probabilities': True,\n",
              " 'ctfidf_model': ClassTfidfTransformer(),\n",
              " 'embedding_model': None,\n",
              " 'hdbscan_model': KMeans(n_clusters=112),\n",
              " 'language': 'english',\n",
              " 'low_memory': False,\n",
              " 'min_topic_size': 30,\n",
              " 'n_gram_range': (1, 3),\n",
              " 'nr_topics': None,\n",
              " 'representation_model': None,\n",
              " 'seed_topic_list': None,\n",
              " 'top_n_words': 10,\n",
              " 'umap_model': UMAP(angular_rp_forest=True, metric='cosine', min_dist=0.0, n_components=5, n_jobs=1, random_state=100, tqdm_kwds={'bar_format': '{desc}: {percentage:3.0f}%| {bar} {n_fmt}/{total_fmt} [{elapsed}]', 'desc': 'Epochs completed', 'disable': True}),\n",
              " 'vectorizer_model': CountVectorizer(ngram_range=(1, 3)),\n",
              " 'verbose': True,\n",
              " 'zeroshot_min_similarity': 0.7,\n",
              " 'zeroshot_topic_list': None}"
            ]
          },
          "metadata": {},
          "execution_count": 28
        }
      ],
      "source": [
        "# Print the parameters used. (For reporting)\n",
        "topic_model.get_params()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G8hwSK98RM9-",
        "outputId": "07a40193-c69a-4356-b1b0-ac721ef78c08"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Done\n"
          ]
        }
      ],
      "source": [
        "tm_params = dict(topic_model.get_params())\n",
        "for key, value in tm_params.items():\n",
        "    tm_params[key]=  str(value)\n",
        "with open(f'{tm_folder_path}/topic_model_params.json', 'w') as f:\n",
        "    json.dump(tm_params, f, ensure_ascii=False, indent=4)\n",
        "    print('Done')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the topic score for each paper and its assigned topic\n",
        "topic_distr, _ = topic_model.approximate_distribution(documents, batch_size=1000)\n",
        "distributions = [distr[topic] if topic != -1 else 0 for topic, distr in zip(topics, topic_distr)]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5v4Wvn3ZAkZP",
        "outputId": "ea347717-a33e-40ad-ad2d-51798cbc5b5f"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 12/12 [00:39<00:00,  3.27s/it]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "topic_distr"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6OTw5go4yVvM",
        "outputId": "7dfa090e-776f-4922-d611-325b1510ccab"
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([[0., 0., 0., ..., 0., 0., 0.],\n",
              "       [0., 0., 0., ..., 0., 0., 0.],\n",
              "       [0., 0., 0., ..., 0., 0., 0.],\n",
              "       ...,\n",
              "       [0., 0., 0., ..., 0., 0., 0.],\n",
              "       [0., 0., 0., ..., 0., 0., 0.],\n",
              "       [1., 0., 0., ..., 0., 0., 0.]])"
            ]
          },
          "metadata": {},
          "execution_count": 36
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 591
        },
        "id": "I7E0_6bv9hBp",
        "outputId": "06faea2f-1c00-4ca7-958b-4dd2dcb10c36"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                    UT                                           Document  \\\n",
              "0  WOS:001333148000001  automated analysis novel object recognition te...   \n",
              "1  WOS:001328968200001  ani-2 enabled open-source protocol estimate li...   \n",
              "2  WOS:001304011000001  accelerated approval cancer drugs united state...   \n",
              "3  WOS:001338099800007  pathological features intrathoracic histiocyti...   \n",
              "4  WOS:001323906200001  classification known unknown study items memor...   \n",
              "\n",
              "   Topic                                      Name  \\\n",
              "0     59      59_speech_language_gaze_participants   \n",
              "1     41       41_protein_receptor_peptide_binding   \n",
              "2     13  13_patients_cancer_chemotherapy_survival   \n",
              "3     51      51_cancer_patients_mutations_genomic   \n",
              "4     18             18_motor_cortex_task_auditory   \n",
              "\n",
              "                                      Representation  \\\n",
              "0  [speech, language, gaze, participants, speaker...   \n",
              "1  [protein, receptor, peptide, binding, proteins...   \n",
              "2  [patients, cancer, chemotherapy, survival, col...   \n",
              "3  [cancer, patients, mutations, genomic, cases, ...   \n",
              "4  [motor, cortex, task, auditory, walking, movem...   \n",
              "\n",
              "                                 Representative_Docs  \\\n",
              "0  [developmental pattern native non-native speec...   \n",
              "1  [structural dynamic insights activation μ-opio...   \n",
              "2  [adjuvant nivolumab plus chemotherapy versus p...   \n",
              "3  [genomic landscape endometrial, ovarian, cervi...   \n",
              "4  [stable sound decoding despite modulated sound...   \n",
              "\n",
              "                                         Top_n_words  Representative_document  \\\n",
              "0  speech - language - gaze - participants - spea...                    False   \n",
              "1  protein - receptor - peptide - binding - prote...                    False   \n",
              "2  patients - cancer - chemotherapy - survival - ...                    False   \n",
              "3  cancer - patients - mutations - genomic - case...                    False   \n",
              "4  motor - cortex - task - auditory - walking - m...                    False   \n",
              "\n",
              "      Score  X_C  level0   cl99  cl-99  \n",
              "0  0.000000   59      59  False  False  \n",
              "1  0.134972   41      41  False  False  \n",
              "2  0.000000   13      13  False  False  \n",
              "3  0.000000   51      51  False  False  \n",
              "4  0.000000   18      18  False  False  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6d702b43-77ab-4c9b-b805-260df8fdfd97\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>UT</th>\n",
              "      <th>Document</th>\n",
              "      <th>Topic</th>\n",
              "      <th>Name</th>\n",
              "      <th>Representation</th>\n",
              "      <th>Representative_Docs</th>\n",
              "      <th>Top_n_words</th>\n",
              "      <th>Representative_document</th>\n",
              "      <th>Score</th>\n",
              "      <th>X_C</th>\n",
              "      <th>level0</th>\n",
              "      <th>cl99</th>\n",
              "      <th>cl-99</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>WOS:001333148000001</td>\n",
              "      <td>automated analysis novel object recognition te...</td>\n",
              "      <td>59</td>\n",
              "      <td>59_speech_language_gaze_participants</td>\n",
              "      <td>[speech, language, gaze, participants, speaker...</td>\n",
              "      <td>[developmental pattern native non-native speec...</td>\n",
              "      <td>speech - language - gaze - participants - spea...</td>\n",
              "      <td>False</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>59</td>\n",
              "      <td>59</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>WOS:001328968200001</td>\n",
              "      <td>ani-2 enabled open-source protocol estimate li...</td>\n",
              "      <td>41</td>\n",
              "      <td>41_protein_receptor_peptide_binding</td>\n",
              "      <td>[protein, receptor, peptide, binding, proteins...</td>\n",
              "      <td>[structural dynamic insights activation μ-opio...</td>\n",
              "      <td>protein - receptor - peptide - binding - prote...</td>\n",
              "      <td>False</td>\n",
              "      <td>0.134972</td>\n",
              "      <td>41</td>\n",
              "      <td>41</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>WOS:001304011000001</td>\n",
              "      <td>accelerated approval cancer drugs united state...</td>\n",
              "      <td>13</td>\n",
              "      <td>13_patients_cancer_chemotherapy_survival</td>\n",
              "      <td>[patients, cancer, chemotherapy, survival, col...</td>\n",
              "      <td>[adjuvant nivolumab plus chemotherapy versus p...</td>\n",
              "      <td>patients - cancer - chemotherapy - survival - ...</td>\n",
              "      <td>False</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>13</td>\n",
              "      <td>13</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>WOS:001338099800007</td>\n",
              "      <td>pathological features intrathoracic histiocyti...</td>\n",
              "      <td>51</td>\n",
              "      <td>51_cancer_patients_mutations_genomic</td>\n",
              "      <td>[cancer, patients, mutations, genomic, cases, ...</td>\n",
              "      <td>[genomic landscape endometrial, ovarian, cervi...</td>\n",
              "      <td>cancer - patients - mutations - genomic - case...</td>\n",
              "      <td>False</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>51</td>\n",
              "      <td>51</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>WOS:001323906200001</td>\n",
              "      <td>classification known unknown study items memor...</td>\n",
              "      <td>18</td>\n",
              "      <td>18_motor_cortex_task_auditory</td>\n",
              "      <td>[motor, cortex, task, auditory, walking, movem...</td>\n",
              "      <td>[stable sound decoding despite modulated sound...</td>\n",
              "      <td>motor - cortex - task - auditory - walking - m...</td>\n",
              "      <td>False</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>18</td>\n",
              "      <td>18</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6d702b43-77ab-4c9b-b805-260df8fdfd97')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-6d702b43-77ab-4c9b-b805-260df8fdfd97 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-6d702b43-77ab-4c9b-b805-260df8fdfd97');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-08105974-da01-43c2-aa41-eb83b4f1769b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-08105974-da01-43c2-aa41-eb83b4f1769b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-08105974-da01-43c2-aa41-eb83b4f1769b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "dataset_clustering_results",
              "summary": "{\n  \"name\": \"dataset_clustering_results\",\n  \"rows\": 11120,\n  \"fields\": [\n    {\n      \"column\": \"UT\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 11120,\n        \"samples\": [\n          \"WOS:001166255900001\",\n          \"WOS:001119133100001\",\n          \"WOS:001287551500001\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Document\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 11106,\n        \"samples\": [\n          \"outskirts abell 1795: probing gas clumping intracluster medium contact. outskirts galaxy clusters host complex interactions intracluster circumcluster media. evolution clusters, ram-pressure stripped gas clumps infalling substructures break uniformity gas distribution, potentially leading observational biases large radii. however, assessing contribution gas clumping poses observational challenges requires robust x-ray measurements background-dominated regime cluster outskirts. aims. main objectives study isolate faint gas clumps field sources diffuse emission abell 1795 galaxy cluster, probe impact observed surface brightness thermodynamic profiles. methods. performed imaging analysis deep chandra acis-i observations abell 1795 cluster outskirts, extending similar 1.5r(200) azimuthal coverage. built 0.7-2.0 kev surface brightness distribution adaptively binned image diffuse emission looked clumps form > +2 sigma surface brightness outliers. classification clump candidates based primarily chandra sdss data. benefiting point source list resolved chandra, extracted thermodynamic profiles intracluster medium associated suzaku xis data r(200) multiple point source clump candidate removal approaches. results. identified 24 clump candidates abell 1795 field, likely associated background objects, including active galactic nuclei, galaxies, clusters groups galaxies, opposed intrinsic gas clumps. sources minimal impact surface brightness thermodynamic profiles cluster emission. correcting clump candidates, measured entropy profile deviates pure gravitational collapse, suggesting complex physics play outskirts, include potential electron-ion non-equilibrium non-thermal pressure support.\",\n          \"impact tetrodotoxin (ttx) gut microbiome juvenile tiger pufferfish, <i>takifugu rubripes</i> tetrodotoxin (ttx) potent neurotoxin accumulates takifugu rubripes, commonly known pufferfish, ingestion ttx-bearing organisms food chain. researchers believe pufferfish use ttx relieve stress, data currently available ttx affects gut microbiota pufferfish. address gap, study aimed investigate administering ttx fish alter gut microbiota overall health salinity conditions, including 30.0 ppt, 8.5 ppt, 1.7 ppt salinity, represent full-strength, isosmotic, low-salinity stress, respectively. analyzed effect ttx ingestion community structure, core microbiome, metabolic capabilities gut microbiome high-throughput sequencing technologies. predominant bacterial taxa gut microbiome firmicutes (21-85%), campilobacterota (2.8-67%), spirochaetota (0.5-14%), proteobacteria (0.7-9.8%), mycoplasma, uncultured arcobacteraceae, brevinema, vibrio, rubritalea, uncultured pirellulaceae core genera. findings indicated impact ttx high-abundance genera 30.0 ppt 8.5 ppt salinity levels negligible, indicating stability resilience ttx ingestion. however, 1.7 ppt, ttx-fed fish showed significant increase uncultured arcobacteraceae. furthermore, analysis ttx-fed fish revealed taxonomic alterations low-abundance taxa, altered predicted functions gut microbiota salinity levels. results suggest ttx administration cause subtle effects metabolic functions gut microbial communities. overall, study provides insights complex relationship ttx-accumulating animal, t. rubripes, gut microbiota.\",\n          \"entanglement-enabled decoherence-free transmission two-color photons single mode fiber demonstrate decoherence-tolerant transmission bell state singlemode fiber (smf) photon frequency degree freedom. end, slightly nondegenerate polarization-entangled singlet, |'f-\\u27e9 = (|hv\\u27e9 - |vh\\u27e9)/ 2, localized smf depolarization channel subject random noise incessant fiber bending. two-photon beats quantum state tomography jointly verify absence collective decoherence, showing |'f-\\u27e9 sought-after one-dimensional decoherence-free subspace (dfs) pertaining smf. efficient splitting combining photon streams dfs help outperform dfs time domain. motivates attempt dfs-enabled fault-tolerant fiber transmission biphoton qubits. two-photon bb84 protocol implemented polarization-maintaining fiber dephasing noise relevant two-dimensional dfs appropriate. low bit error rate 5.4% achieved encoding one-qubit information biphoton state spite significant polarization fluctuation. scalable frequency-based dfs natural affinity wavelength division multiplexing fiber communication design extensible multi-particle entanglement.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Topic\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 30,\n        \"min\": 0,\n        \"max\": 111,\n        \"num_unique_values\": 112,\n        \"samples\": [\n          70,\n          26,\n          18\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 112,\n        \"samples\": [\n          \"70_robot_control_robots_proposed\",\n          \"26_mice_muscle_senescence_expression\",\n          \"18_motor_cortex_task_auditory\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representation\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_Docs\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Top_n_words\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 112,\n        \"samples\": [\n          \"robot - control - robots - proposed - method - motion - controller - tracking - trajectory - learning\",\n          \"mice - muscle - senescence - expression - cells - skeletal - mitochondrial - metabolism - metabolic - glucose\",\n          \"motor - cortex - task - auditory - walking - movement - brain - learning - neural - activity\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Representative_document\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          true,\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Score\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0.4215785769906425,\n        \"min\": 0.0,\n        \"max\": 1.0,\n        \"num_unique_values\": 1589,\n        \"samples\": [\n          0.9017271449581243,\n          0.32495595858398113\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"X_C\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 30,\n        \"min\": 0,\n        \"max\": 111,\n        \"num_unique_values\": 112,\n        \"samples\": [\n          70,\n          26\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"level0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 30,\n        \"min\": 0,\n        \"max\": 111,\n        \"num_unique_values\": 112,\n        \"samples\": [\n          70,\n          26\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cl99\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"cl-99\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          false\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 31
        }
      ],
      "source": [
        "# Document information. Including the topic assignation\n",
        "dataset_clustering_results = topic_model.get_document_info(documents, df = corpus, metadata={\"Score\": distributions})\n",
        "\n",
        "# Standar format for report analysis\n",
        "dataset_clustering_results = dataset_clustering_results.drop(columns=['text'])\n",
        "dataset_clustering_results['X_E'] = dataset_clustering_results['Score']\n",
        "dataset_clustering_results['X_C'] = dataset_clustering_results['Topic'] + 1\n",
        "dataset_clustering_results['level0'] = dataset_clustering_results['Topic'] + 1\n",
        "dataset_clustering_results['cl99'] = False\n",
        "dataset_clustering_results['cl-99'] = False\n",
        "dataset_clustering_results.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {
        "id": "eyQDhrjyc6Oh"
      },
      "outputs": [],
      "source": [
        "# Save the dataframe\n",
        "dataset_clustering_results.to_csv(f'{tm_folder_path}/dataset_minimal.csv', index=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1RYcsXanZvrr",
        "outputId": "d625d86f-5897-4d75-8b32-54f2dd258ad3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "2024-11-02 06:45:14,320 - BERTopic - WARNING: When you use `pickle` to save/load a BERTopic model,please make sure that the environments in which you saveand load the model are **exactly** the same. The version of BERTopic,its dependencies, and python need to remain the same.\n"
          ]
        }
      ],
      "source": [
        "# Save the topic model\n",
        "topic_model.save(f'{tm_folder_path}/topic_model_object.pck')"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "SX1EBSOY3Pz1"
      }
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "authorship_tag": "ABX9TyM4oE6dpT6UwjIg42kzzsWo",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}