{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/cristianmejia00/clustering/blob/main/Topic_Models_using_BERTopic_TOPIC_MODEL_20241101.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Zj3lYckpO2Yn"
   },
   "source": [
    "# Topic Modeling with BERTopic\n",
    "\n",
    "üî¥ copied from the [Kubota Colab](https://colab.research.google.com/drive/1YsDp5_qGXGJKsEXsS8DO8CA_lqZc6EpA).  \n",
    "\n",
    "`Topic Models` are methods to automatically organize a corpus of text into topics.\n",
    "\n",
    "Topic Model process:\n",
    "1. Data preparation\n",
    "2. Tranform text to numeric vectors\n",
    "3. Multidimensionality reduction\n",
    "4. Clustering\n",
    "5. Topic analysis\n",
    "6. Cluster assignation\n",
    "\n",
    "\n",
    "This notebook uses the library `BERTopic` which is a one-stop solution for topic modeling including handy functions for plotting and analysis. However, BERTopic does not have a function to extract the X and Y coords from UMAP. If we need the coordinates then use the notebooks `Topic_Models_using_Transformers` instead. In any other situation, when a quick analysis is needed this notebook may be better.\n",
    "\n",
    "This notebook is also the one needed for the heatmap codes included in this folder.\n",
    "\n",
    "`BERTopic` is Python library that handles steps 2 to 6.\n",
    "BERT topic models use the transformer architechture to generate the embeds (i.e. the vector or numeric representation of words) and are currently the state-of-the-art method for vectorization.\n",
    "\n",
    "This notebook shows how to use it.\n",
    "\n",
    "---\n",
    "Reading:\n",
    "[Topic Modeling with Deep Learning Using Python BERTopic](https://medium.com/grabngoinfo/topic-modeling-with-deep-learning-using-python-bertopic-cf91f5676504)\n",
    "[Advanced Topic Modeling with BERTopic](https://www.pinecone.io/learn/bertopic/)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WlufmOowbHSF"
   },
   "source": [
    "# Requirements"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "kFQU-BC5ReYd"
   },
   "source": [
    "## Packages installation and initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "gMlZ_7DkOxGG"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/cristian/Desktop/GitHub/clustering/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "import pickle\n",
    "from datetime import date\n",
    "from itertools import compress\n",
    "from bertopic import BERTopic\n",
    "from umap import UMAP\n",
    "from sklearn.cluster import KMeans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dDm5-r5oR98q"
   },
   "source": [
    "# üî¥ Input files and options\n",
    "\n",
    "Go to your Google Drive and create a folder in the root directory. We are going to save all related data in that directory.\n",
    "Upload the dataset of news into the above folder.\n",
    "- The dataset should be a `.csv` file.\n",
    "- Every row in the dataset is a document\n",
    "- It can any kind of columns. Some columns must contain the text we want to analyze. For example, a dataset of academic articles may contain a \"Title\" and/or \"Abstract\" column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "5ICq5zQbSJa8"
   },
   "outputs": [],
   "source": [
    "# The bibliometrics folder\n",
    "# Colab\n",
    "ROOT_FOLDER_PATH = \"drive/MyDrive/Bibliometrics_Drive\"\n",
    "\n",
    "# Mac\n",
    "ROOT_FOLDER_PATH = \"/Users/cristian/Library/CloudStorage/GoogleDrive-cristianmejia00@gmail.com/My Drive/Bibliometrics_Drive\"\n",
    "\n",
    "# Change to the name of the folder where the dataset is uploaded inside the above folder\n",
    "project_folder = 'Q339_igem'\n",
    "\n",
    "# Analysis ID\n",
    "analysis_id = 'a01_tm__f01_e01__hdbs'\n",
    "\n",
    "# Filtered label\n",
    "settings_directive = \"settings_analysis_directive_2025-08-04-16-40.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "QpoR8P3O0ACf"
   },
   "outputs": [],
   "source": [
    "# Read settings\n",
    "with open(f'{ROOT_FOLDER_PATH}/{project_folder}/{analysis_id}/{settings_directive}', 'r') as file:\n",
    "    settings = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "COh3zc3b2oAn"
   },
   "outputs": [],
   "source": [
    "# Input dataset\n",
    "dataset_file_path = f\"{ROOT_FOLDER_PATH}/{settings['metadata']['project_folder']}/{settings['metadata']['filtered_folder']}/dataset_raw_cleaned.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "biEFdxrm0WrO"
   },
   "outputs": [],
   "source": [
    "# Function to save files\n",
    "def save_as_csv(df, save_name_without_extension, with_index):\n",
    "    \"usage: `save_as_csv(dataframe, 'filename')`\"\n",
    "    df.to_csv(f\"{ROOT_FOLDER_PATH}/{save_name_without_extension}.csv\", index=with_index)\n",
    "    print(\"===\\nSaved: \", f\"{ROOT_FOLDER_PATH}/{save_name_without_extension}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "id": "Ff7hQXE1dYDw"
   },
   "outputs": [],
   "source": [
    "# prompt: a function to save object to a pickle file\n",
    "def save_object_as_pickle(obj, filename):\n",
    "  \"\"\"\n",
    "  Saves an object as a pickle file.\n",
    "\n",
    "  Args:\n",
    "      obj: The object to be saved.\n",
    "      filename: The filename of the pickle file.\n",
    "  \"\"\"\n",
    "  with open(filename, \"wb\") as f:\n",
    "    pickle.dump(obj, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "iljQe0_xb2FU"
   },
   "outputs": [],
   "source": [
    "# prompt: a function to load pickle object given a path\n",
    "def load_pickle(path):\n",
    "    with open(path, 'rb') as f:\n",
    "        return pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 655
    },
    "id": "Tob-4BIaUbZ9",
    "outputId": "d48eac6d-2af8-4991-af19-5bda2386e6bc"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4199, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X_N</th>\n",
       "      <th>uuid</th>\n",
       "      <th>UT</th>\n",
       "      <th>PY</th>\n",
       "      <th>TI</th>\n",
       "      <th>AB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>f399a9de-8ec5-4ace-b6f3-b9137f3107eb</td>\n",
       "      <td>173</td>\n",
       "      <td>2009</td>\n",
       "      <td>Ethanol? Whey not!</td>\n",
       "      <td>Cheese whey is classified as a special waste f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>9a1131bd-91da-4871-9bad-d809761bc77e</td>\n",
       "      <td>174</td>\n",
       "      <td>2009</td>\n",
       "      <td>Bac-man: sequestering cadmium into Bacillus sp...</td>\n",
       "      <td>Cadmium contamination can be a serious problem...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>f866ee62-25b0-4cde-b895-45009fa2b6fe</td>\n",
       "      <td>175</td>\n",
       "      <td>2009</td>\n",
       "      <td>Bacterial Relay Race</td>\n",
       "      <td>In our project, we aim at creating a cell-to-c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>32a634be-6e1d-4d11-bace-8a31f9a3a88a</td>\n",
       "      <td>176</td>\n",
       "      <td>2009</td>\n",
       "      <td>E. coli Automatic Directed Evolution Machine: ...</td>\n",
       "      <td>Evolution is powerful enough to create everyth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>63d73d3a-b239-4624-a2bb-2fb6d068294e</td>\n",
       "      <td>177</td>\n",
       "      <td>2009</td>\n",
       "      <td>BacInVader √ê a new system for cancer genetic t...</td>\n",
       "      <td>The main aim of our project is to design a mod...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   X_N                                  uuid   UT    PY  \\\n",
       "0    1  f399a9de-8ec5-4ace-b6f3-b9137f3107eb  173  2009   \n",
       "1    2  9a1131bd-91da-4871-9bad-d809761bc77e  174  2009   \n",
       "2    3  f866ee62-25b0-4cde-b895-45009fa2b6fe  175  2009   \n",
       "3    4  32a634be-6e1d-4d11-bace-8a31f9a3a88a  176  2009   \n",
       "4    5  63d73d3a-b239-4624-a2bb-2fb6d068294e  177  2009   \n",
       "\n",
       "                                                  TI  \\\n",
       "0                                 Ethanol? Whey not!   \n",
       "1  Bac-man: sequestering cadmium into Bacillus sp...   \n",
       "2                               Bacterial Relay Race   \n",
       "3  E. coli Automatic Directed Evolution Machine: ...   \n",
       "4  BacInVader √ê a new system for cancer genetic t...   \n",
       "\n",
       "                                                  AB  \n",
       "0  Cheese whey is classified as a special waste f...  \n",
       "1  Cadmium contamination can be a serious problem...  \n",
       "2  In our project, we aim at creating a cell-to-c...  \n",
       "3  Evolution is powerful enough to create everyth...  \n",
       "4  The main aim of our project is to design a mod...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Open the data file\n",
    "df = pd.read_csv(f\"{dataset_file_path}\", encoding='latin-1', usecols=['X_N', 'uuid', 'UT', 'PY', 'TI', 'AB'])\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "9UrPiLepN1s9"
   },
   "source": [
    "\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "eK-U-Sl8fc-m"
   },
   "source": [
    "## PART 2: Topic Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "FhwGnBjFQT4x"
   },
   "outputs": [],
   "source": [
    "# bibliometrics_folder\n",
    "# project_folder\n",
    "# project_name_suffix\n",
    "# ROOT_FOLDER_PATH = f\"drive/MyDrive/{bibliometrics_folder}\"\n",
    "\n",
    "#############################################################\n",
    "# Embeddings folder\n",
    "embeddings_folder_name = settings['tmo']['embeds_folder']\n",
    "\n",
    "# Which column has the year of the documents?\n",
    "my_year = settings['tmo']['year_column']\n",
    "\n",
    "# Number of topics. Select the number of topics to extract.\n",
    "# Choose 0, for automatic detection.\n",
    "n_topics = 71#settings['tmo']['n_topics']\n",
    "\n",
    "# Minimum number of documents per topic\n",
    "min_topic_size = settings['tmo']['min_topic_size']\n",
    "\n",
    "# Threshold for others\n",
    "# This is the threshold for the topics that will be considered as \"others\"\n",
    "# Topics with a cumulative percentage of documents below this threshold will be grouped into an \"others\" topic.\n",
    "# For example, if set to 0.9, topics that together account for 90% of the documents will be kept, and the rest will be grouped into \"others\".\n",
    "others_threshold = 0.99#settings['tmo']['others_threshold']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "9RYSioo2cxxC"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embeddings shape: (4199, 384)\n"
     ]
    }
   ],
   "source": [
    "# Get the embeddings back.\n",
    "embeddings = load_pickle(f\"{ROOT_FOLDER_PATH}/{project_folder}/{settings['metadata']['filtered_folder']}/{embeddings_folder_name}/embeddings.pck\")\n",
    "corpus =     pd.read_csv(f\"{ROOT_FOLDER_PATH}/{project_folder}/{settings['metadata']['filtered_folder']}/{embeddings_folder_name}/corpus.csv\").reset_index(drop=True)\n",
    "documents = corpus.text.to_list()\n",
    "print(f\"Embeddings shape: {embeddings['embeddings'].shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "S7LeuG_7S4Db",
    "outputId": "4e27c106-2e0c-4630-a741-2fbd5e5512b0"
   },
   "outputs": [],
   "source": [
    "# verify that the number of embeddings matches the number of documents\n",
    "assert(len(embeddings['embeddings']) == len(documents))\n",
    "\n",
    "# verify that embedding and documents are in the same order as the original dataframe\n",
    "assert(corpus['uuid'].tolist() == embeddings['embeddings_ids'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Topic Modeling with BERTopic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "id": "n6qUvhGGcPZo"
   },
   "outputs": [],
   "source": [
    "from hdbscan.hdbscan_ import HDBSCAN\n",
    "# Execute the topic model.\n",
    "# I suggest changing the values marked with #<---\n",
    "# The others are the default values and they'll work fine in most cases.\n",
    "# This will take several minutes to finish.\n",
    "\n",
    "# Initiate UMAP\n",
    "umap_model = UMAP(n_neighbors=15,\n",
    "                  n_components=5,\n",
    "                  min_dist=0.0,\n",
    "                  metric='cosine',\n",
    "                  random_state=100)\n",
    "\n",
    "if n_topics == 0:\n",
    "  # Initiate topic model with HDBScan (Automatic topic selection)\n",
    "  topic_model_params = HDBSCAN(min_cluster_size=min_topic_size,\n",
    "                               metric='euclidean',\n",
    "                               cluster_selection_method='eom',\n",
    "                               prediction_data=True)\n",
    "else:\n",
    "  # Initiate topic model with K-means (Manual topic selection)\n",
    "  topic_model_params = KMeans(n_clusters = n_topics)\n",
    "\n",
    "# Initiate BERTopic\n",
    "topic_model = BERTopic(umap_model = umap_model,\n",
    "                       hdbscan_model = topic_model_params,\n",
    "                       min_topic_size=min_topic_size,\n",
    "                       #nr_topics=15,          #<--- Footnote 1\n",
    "                       n_gram_range=(1,3),\n",
    "                       language='english',\n",
    "                       calculate_probabilities=True,\n",
    "                       verbose=True)\n",
    "\n",
    "\n",
    "\n",
    "# Footnote 1: This controls the number of topics we want AFTER clustering.\n",
    "# Add a hashtag at the beggining to use the number of topics returned by the topic model.\n",
    "# When using HDBScan nr_topics will be obtained after orphans removal, and there is no warranty that `nr_topics < HDBScan topics`.\n",
    "# thus, with HDBScan `nr_topics` means N topics OR LESS.\n",
    "# When using KMeans nr_topics can be used to further reduce the number of topics.\n",
    "# We use the topics as returned by the topic model. So we do not need to activate it here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2026-01-19 15:16:47,276 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm\n",
      "2026-01-19 15:16:59,503 - BERTopic - Dimensionality - Completed ‚úì\n",
      "2026-01-19 15:16:59,504 - BERTopic - Cluster - Start clustering the reduced embeddings\n",
      "2026-01-19 15:16:59,607 - BERTopic - Cluster - Completed ‚úì\n",
      "2026-01-19 15:16:59,612 - BERTopic - Representation - Fine-tuning topics using representation models.\n",
      "2026-01-19 15:17:01,646 - BERTopic - Representation - Completed ‚úì\n"
     ]
    }
   ],
   "source": [
    "# Compute topic model\n",
    "topics, probabilities = topic_model.fit_transform(documents, embeddings['embeddings'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 424
    },
    "id": "67pgsMtwsFmH",
    "outputId": "82c86e95-6be7-4832-a076-94d8d9dcd8ae"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "      <td>0_the_of_and_production</td>\n",
       "      <td>[the, of, and, production, to, in, yeast, for,...</td>\n",
       "      <td>[Fruit wine brewing Plus The high content of h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>122</td>\n",
       "      <td>1_cancer_cells_the_to</td>\n",
       "      <td>[cancer, cells, the, to, of, and, in, tumor, t...</td>\n",
       "      <td>[Project CARKinos  a better way to treat Lung ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>121</td>\n",
       "      <td>2_the_to_nitrogen_and</td>\n",
       "      <td>[the, to, nitrogen, and, of, water, in, is, al...</td>\n",
       "      <td>[Smart Algicidal Bacteria to Control Water Blo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>120</td>\n",
       "      <td>3_light_the_of_and</td>\n",
       "      <td>[light, the, of, and, to, in, we, cell, is, co...</td>\n",
       "      <td>[Lights Camera Flip Engineering a LightActivat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>115</td>\n",
       "      <td>4_metal_heavy_heavy metal_metals</td>\n",
       "      <td>[metal, heavy, heavy metal, metals, the, and, ...</td>\n",
       "      <td>[Nanocrystal Ecoli Flocculation Union Heavy me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>66</td>\n",
       "      <td>17</td>\n",
       "      <td>66_clock_melatonin_circadian_the</td>\n",
       "      <td>[clock, melatonin, circadian, the, insomnia, o...</td>\n",
       "      <td>[Transplanting KaiABC Reduce your jet lag with...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>67</td>\n",
       "      <td>14</td>\n",
       "      <td>67_crispr_to_gene_drive</td>\n",
       "      <td>[crispr, to, gene, drive, in, system, we, of, ...</td>\n",
       "      <td>[Bioengineering a mechanism to override plasmi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>68</td>\n",
       "      <td>13</td>\n",
       "      <td>68_pfas_pfoa_per and_of pfas</td>\n",
       "      <td>[pfas, pfoa, per and, of pfas, substances, and...</td>\n",
       "      <td>[Detection and Degradation of Perfluoroalkyl S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>69</td>\n",
       "      <td>11</td>\n",
       "      <td>69_chromium_hexavalent chromium_hexavalent_oxy...</td>\n",
       "      <td>[chromium, hexavalent chromium, hexavalent, ox...</td>\n",
       "      <td>[OxygenMAX Previous work have shown that the e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70</td>\n",
       "      <td>10</td>\n",
       "      <td>70_nan_sententiae_sit_est</td>\n",
       "      <td>[nan, sententiae, sit, est, sententiae est, ip...</td>\n",
       "      <td>[CerviCare nan, Run Sperm Run nan, Lorem ipsum...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows √ó 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic  Count                                               Name  \\\n",
       "0       0    150                            0_the_of_and_production   \n",
       "1       1    122                              1_cancer_cells_the_to   \n",
       "2       2    121                              2_the_to_nitrogen_and   \n",
       "3       3    120                                 3_light_the_of_and   \n",
       "4       4    115                   4_metal_heavy_heavy metal_metals   \n",
       "..    ...    ...                                                ...   \n",
       "66     66     17                   66_clock_melatonin_circadian_the   \n",
       "67     67     14                            67_crispr_to_gene_drive   \n",
       "68     68     13                       68_pfas_pfoa_per and_of pfas   \n",
       "69     69     11  69_chromium_hexavalent chromium_hexavalent_oxy...   \n",
       "70     70     10                          70_nan_sententiae_sit_est   \n",
       "\n",
       "                                       Representation  \\\n",
       "0   [the, of, and, production, to, in, yeast, for,...   \n",
       "1   [cancer, cells, the, to, of, and, in, tumor, t...   \n",
       "2   [the, to, nitrogen, and, of, water, in, is, al...   \n",
       "3   [light, the, of, and, to, in, we, cell, is, co...   \n",
       "4   [metal, heavy, heavy metal, metals, the, and, ...   \n",
       "..                                                ...   \n",
       "66  [clock, melatonin, circadian, the, insomnia, o...   \n",
       "67  [crispr, to, gene, drive, in, system, we, of, ...   \n",
       "68  [pfas, pfoa, per and, of pfas, substances, and...   \n",
       "69  [chromium, hexavalent chromium, hexavalent, ox...   \n",
       "70  [nan, sententiae, sit, est, sententiae est, ip...   \n",
       "\n",
       "                                  Representative_Docs  \n",
       "0   [Fruit wine brewing Plus The high content of h...  \n",
       "1   [Project CARKinos  a better way to treat Lung ...  \n",
       "2   [Smart Algicidal Bacteria to Control Water Blo...  \n",
       "3   [Lights Camera Flip Engineering a LightActivat...  \n",
       "4   [Nanocrystal Ecoli Flocculation Union Heavy me...  \n",
       "..                                                ...  \n",
       "66  [Transplanting KaiABC Reduce your jet lag with...  \n",
       "67  [Bioengineering a mechanism to override plasmi...  \n",
       "68  [Detection and Degradation of Perfluoroalkyl S...  \n",
       "69  [OxygenMAX Previous work have shown that the e...  \n",
       "70  [CerviCare nan, Run Sperm Run nan, Lorem ipsum...  \n",
       "\n",
       "[71 rows x 5 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the list of topics\n",
    "# Topic = the topic number. From the largest topic.\n",
    "#         \"-1\" is the generic topic. Genericr keywords are aggegrated here.\n",
    "# Count = Documents assigned to this topic\n",
    "# Name = Top 4 words of the topic based on probability\n",
    "# Representation = The list of words representing this topic\n",
    "# Representative_Docs = Documents assigned to this topic\n",
    "tm_summary = topic_model.get_topic_info()\n",
    "tm_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "id": "k9mN1tgZQupy"
   },
   "outputs": [],
   "source": [
    "# Save the topic model assets\n",
    "tm_folder_path = f'{ROOT_FOLDER_PATH}/{project_folder}/{settings[\"metadata\"][\"analysis_id\"]}'\n",
    "\n",
    "if not os.path.exists(tm_folder_path):\n",
    "  !mkdir $tm_folder_path\n",
    "\n",
    "tm_summary.to_csv(f'{tm_folder_path}/topic_model_info.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "WuE0vkbp3fVL",
    "outputId": "aa8ebf45-cdd3-4161-9ac4-1343c91f47da"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of topics found\n",
    "found_topics = max(tm_summary.Topic) + 1\n",
    "found_topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KjKUcNMdwFHQ",
    "outputId": "b3549b7c-4c8a-4b82-8d87-6a46774e4694"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Confirm all documents are assigned\n",
    "sum(tm_summary.Count) == len(corpus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "PU99D2xhxoOY",
    "outputId": "4e90e928-9bbf-40c3-c3b5-a1a50abe82f0"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Fruit wine brewing Plus The high content of higher alcohol in beverages can easily lead to symptoms such as headache which is the main reason for the slow drunkenness This project finely regulates the metabolic pathways of higher alcohols in Saccharomyces cerevisiae through genetic engineering to reduce the production of higher alcohols First the intracellular homologous recombination technology was used to knock out the branchedchain amino acid aminotransferase encoded by the BAT gene and at the same time the alcohol acetyltransferase ATF gene was used to replace the position of the BAT gene Secondly the ATF gene is placed on the expression vector of Saccharomyces cerevisiae and then transferred to Saccharomyces cerevisiae to achieve the purpose of reducing the level of higher alcohol in Saccharomyces cerevisiae This project makes the brewed beverage have less high alcohol content so that the drinker is not uncomfortable in the head',\n",
       " 'Construction of a Saccharomyces cerevisiae cell factory for high carotene production carotene is a terpenoid compound with important medical food and industrial values Traditional physical extraction methods are timeconsuming and cumbersome and the cultivation of carotenecontaining crops requires geographical location and growth cycle Chemical synthesis of carotene may be toxic limiting its application  Microbial fermentation for the synthesis of carotene is currently the most promising and promising synthetic method Saccharomyces cerevisiae a recognized foodsafe grade microorganism with the advantages of easy molecular manipulation rapid growth and good tolerance has been adapted for the production of various terpenoids including carotene In this study we targeted the carotene synthesis pathway optimized the synthesis pathway genes arranged and combined the candidate gene species sources constructed the corresponding strains tested the carotene yield of the resulting strains and obtained the optimal gene combinations for the construction of a highyielding carotene brewers yeast to assist in the commercial production of carotene',\n",
       " 'Construction of organismextrinsic synthetic pathways for the biosynthesis of beneficial natural compounds Team Minnesota aims to change the paradigm regarding the synthesis of natural products  Rather than depending upon slow and expensive chemical synthesis our team has developed two cohesive platforms utilizing the BioBrick strategy and synthetic biology to produce compounds for public health and nutrition using industriallyrelevant microorganisms First we constructed and optimized designer pathways using the BioBrick platform for the production of a suite of sunscreenlike compounds that inhibit the effects of ultraviolet radiation which we hope to incorporate into bacteria found on the skin microbiome for prolong ultraviolet protection  We also developed a novel and modular BioBrick backbone for expression in Saccharomyces cerevisiae  For demonstration we constructed a caffeine production pathway in this backbone generating a yeast strain which produces caffeine  Both of these projects gain impact from their synergistic application of synthetic biology and bioengineering for products that apply to realworld situations for researchers and the general public']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the top 10 documents for a topic\n",
    "topic_model.get_representative_docs(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "JQKxUqHN9RT7",
    "outputId": "c572d73b-fc91-4c84-dc79-8f047b046aa2"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'calculate_probabilities': True,\n",
       " 'ctfidf_model': ClassTfidfTransformer(),\n",
       " 'embedding_model': None,\n",
       " 'hdbscan_model': KMeans(n_clusters=71),\n",
       " 'language': 'english',\n",
       " 'low_memory': False,\n",
       " 'min_topic_size': 5,\n",
       " 'n_gram_range': (1, 3),\n",
       " 'nr_topics': None,\n",
       " 'representation_model': None,\n",
       " 'seed_topic_list': None,\n",
       " 'top_n_words': 10,\n",
       " 'umap_model': UMAP(angular_rp_forest=True, metric='cosine', min_dist=0.0, n_components=5, n_jobs=1, random_state=100, tqdm_kwds={'bar_format': '{desc}: {percentage:3.0f}%| {bar} {n_fmt}/{total_fmt} [{elapsed}]', 'desc': 'Epochs completed', 'disable': True}),\n",
       " 'vectorizer_model': CountVectorizer(ngram_range=(1, 3)),\n",
       " 'verbose': True,\n",
       " 'zeroshot_min_similarity': 0.7,\n",
       " 'zeroshot_topic_list': None}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Print the parameters used. (For reporting)\n",
    "topic_model.get_params()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "G8hwSK98RM9-",
    "outputId": "07a40193-c69a-4356-b1b0-ac721ef78c08"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "tm_params = dict(topic_model.get_params())\n",
    "for key, value in tm_params.items():\n",
    "    tm_params[key]=  str(value)\n",
    "with open(f'{tm_folder_path}/topic_model_params.json', 'w') as f:\n",
    "    json.dump(tm_params, f, ensure_ascii=False, indent=4)\n",
    "    print('Done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Centroids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.metrics.pairwise import cosine_distances\n",
    "\n",
    "def calculate_centroid_scores(embeddings, topics, distance_metric='euclidean'):\n",
    "    \"\"\"\n",
    "    Vectorized calculation of distance from topic centroids.\n",
    "    Returns normalized scores (0-1) where 0 is the centroid and 1 is far.\n",
    "    \"\"\"\n",
    "    embeddings = np.array(embeddings)\n",
    "    topics = np.array(topics)\n",
    "    unique_topics = np.unique(topics)\n",
    "    \n",
    "    scores = np.zeros(len(embeddings))\n",
    "    \n",
    "    # Calculate centroids and distances\n",
    "    for topic in unique_topics:\n",
    "        if topic == -1: continue # Skip outliers for centroid calculation\n",
    "        \n",
    "        mask = topics == topic\n",
    "        topic_embeds = embeddings[mask]\n",
    "        centroid = np.mean(topic_embeds, axis=0)\n",
    "        \n",
    "        if distance_metric == 'euclidean':\n",
    "            dists = np.linalg.norm(topic_embeds - centroid, axis=1)\n",
    "        else:\n",
    "            dists = cosine_distances([centroid], topic_embeds)[0]\n",
    "            \n",
    "        # Normalize 0-1 within topic\n",
    "        if len(dists) > 1 and dists.max() > dists.min():\n",
    "            scores[mask] = (dists - dists.min()) / (dists.max() - dists.min())\n",
    "        else:\n",
    "            scores[mask] = 0 # Single document or identical distances\n",
    "            \n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usage with normalization:\n",
    "scores = calculate_centroid_scores(embeddings['embeddings'], topics)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>UT</th>\n",
       "      <th>uuid</th>\n",
       "      <th>Document</th>\n",
       "      <th>Topic</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>Representative_Docs</th>\n",
       "      <th>Top_n_words</th>\n",
       "      <th>Representative_document</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ethanol Whey not Cheese whey is classified as ...</td>\n",
       "      <td>173</td>\n",
       "      <td>791aefbd-0fde-417e-8e1b-2f7210647032</td>\n",
       "      <td>Ethanol Whey not Cheese whey is classified as ...</td>\n",
       "      <td>27</td>\n",
       "      <td>27_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, production, in, lactose, mi...</td>\n",
       "      <td>[The Reform of Two Strains in Yogurt Since lot...</td>\n",
       "      <td>of - the - and - to - production - in - lactos...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacman sequestering cadmium into Bacillus spor...</td>\n",
       "      <td>174</td>\n",
       "      <td>84d56c08-f62f-4bfd-986c-71a44c1af2ed</td>\n",
       "      <td>Bacman sequestering cadmium into Bacillus spor...</td>\n",
       "      <td>4</td>\n",
       "      <td>4_metal_heavy_heavy metal_metals</td>\n",
       "      <td>[metal, heavy, heavy metal, metals, the, and, ...</td>\n",
       "      <td>[Nanocrystal Ecoli Flocculation Union Heavy me...</td>\n",
       "      <td>metal - heavy - heavy metal - metals - the - a...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacterial Relay Race In our project we aim at ...</td>\n",
       "      <td>175</td>\n",
       "      <td>8a69e820-ef66-4438-8ed2-08943ab2c6b1</td>\n",
       "      <td>Bacterial Relay Race In our project we aim at ...</td>\n",
       "      <td>25</td>\n",
       "      <td>25_the_of_to_in</td>\n",
       "      <td>[the, of, to, in, and, quorum, system, we, quo...</td>\n",
       "      <td>[A multilayer signalprocessing system based on...</td>\n",
       "      <td>the - of - to - in - and - quorum - system - w...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E coli Automatic Directed Evolution Machine a ...</td>\n",
       "      <td>176</td>\n",
       "      <td>572f8dd4-bb30-4d59-a319-2a5e650cce90</td>\n",
       "      <td>E coli Automatic Directed Evolution Machine a ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, for, in, evolution, synthet...</td>\n",
       "      <td>[rEvolver  Vivo La Evolution Naturallyoccurrin...</td>\n",
       "      <td>of - the - and - to - for - in - evolution - s...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BacInVader  a new system for cancer genetic th...</td>\n",
       "      <td>177</td>\n",
       "      <td>be21c9a1-0aac-4e87-86f5-f6023d374461</td>\n",
       "      <td>BacInVader  a new system for cancer genetic th...</td>\n",
       "      <td>10</td>\n",
       "      <td>10_cancer_tumor_cells_the</td>\n",
       "      <td>[cancer, tumor, cells, the, to, and, of, thera...</td>\n",
       "      <td>[BLAST  Bifidobacterium Longum induced Apoptos...</td>\n",
       "      <td>cancer - tumor - cells - the - to - and - of -...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                text   UT  \\\n",
       "0  Ethanol Whey not Cheese whey is classified as ...  173   \n",
       "1  Bacman sequestering cadmium into Bacillus spor...  174   \n",
       "2  Bacterial Relay Race In our project we aim at ...  175   \n",
       "3  E coli Automatic Directed Evolution Machine a ...  176   \n",
       "4  BacInVader  a new system for cancer genetic th...  177   \n",
       "\n",
       "                                   uuid  \\\n",
       "0  791aefbd-0fde-417e-8e1b-2f7210647032   \n",
       "1  84d56c08-f62f-4bfd-986c-71a44c1af2ed   \n",
       "2  8a69e820-ef66-4438-8ed2-08943ab2c6b1   \n",
       "3  572f8dd4-bb30-4d59-a319-2a5e650cce90   \n",
       "4  be21c9a1-0aac-4e87-86f5-f6023d374461   \n",
       "\n",
       "                                            Document  Topic  \\\n",
       "0  Ethanol Whey not Cheese whey is classified as ...     27   \n",
       "1  Bacman sequestering cadmium into Bacillus spor...      4   \n",
       "2  Bacterial Relay Race In our project we aim at ...     25   \n",
       "3  E coli Automatic Directed Evolution Machine a ...      6   \n",
       "4  BacInVader  a new system for cancer genetic th...     10   \n",
       "\n",
       "                               Name  \\\n",
       "0                  27_of_the_and_to   \n",
       "1  4_metal_heavy_heavy metal_metals   \n",
       "2                   25_the_of_to_in   \n",
       "3                   6_of_the_and_to   \n",
       "4         10_cancer_tumor_cells_the   \n",
       "\n",
       "                                      Representation  \\\n",
       "0  [of, the, and, to, production, in, lactose, mi...   \n",
       "1  [metal, heavy, heavy metal, metals, the, and, ...   \n",
       "2  [the, of, to, in, and, quorum, system, we, quo...   \n",
       "3  [of, the, and, to, for, in, evolution, synthet...   \n",
       "4  [cancer, tumor, cells, the, to, and, of, thera...   \n",
       "\n",
       "                                 Representative_Docs  \\\n",
       "0  [The Reform of Two Strains in Yogurt Since lot...   \n",
       "1  [Nanocrystal Ecoli Flocculation Union Heavy me...   \n",
       "2  [A multilayer signalprocessing system based on...   \n",
       "3  [rEvolver  Vivo La Evolution Naturallyoccurrin...   \n",
       "4  [BLAST  Bifidobacterium Longum induced Apoptos...   \n",
       "\n",
       "                                         Top_n_words  Representative_document  \n",
       "0  of - the - and - to - production - in - lactos...                    False  \n",
       "1  metal - heavy - heavy metal - metals - the - a...                    False  \n",
       "2  the - of - to - in - and - quorum - system - w...                    False  \n",
       "3  of - the - and - to - for - in - evolution - s...                    False  \n",
       "4  cancer - tumor - cells - the - to - and - of -...                    False  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_clustering_results = topic_model.get_document_info(documents, df = corpus)\n",
    "dataset_clustering_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating 2D coordinates for visualization...\n"
     ]
    }
   ],
   "source": [
    "from sklearn.manifold import TSNE\n",
    "\n",
    "# 6. VISUALIZATION (Switching to t-SNE for the \"Map\" look)\n",
    "print(\"Generating 2D coordinates for visualization...\")\n",
    "\n",
    "# Filter out outliers (-1) to keep the visualization clean\n",
    "valid_indices = dataset_clustering_results['Topic'] != -1\n",
    "valid_embeddings = np.array(embeddings['embeddings'])[valid_indices]\n",
    "df_clean = dataset_clustering_results[valid_indices].reset_index(drop=True)\n",
    "\n",
    "# --- KEY CHANGE: Use t-SNE instead of UMAP for the layout ---\n",
    "# metric='cosine' is better for embeddings\n",
    "# init='pca' is CRITICAL: it gives you the \"ball\" shape instead of random clusters\n",
    "# perplexity=30-50: balances local vs global structure\n",
    "tsne_model = TSNE(\n",
    "    n_components=2, \n",
    "    perplexity=50,        # Higher values = more global structure\n",
    "    learning_rate='auto',\n",
    "    init='pca',           # This creates the global \"continent\" shape\n",
    "    #n_iter=1000, \n",
    "    metric='cosine', \n",
    "    random_state=42, \n",
    "    n_jobs=-1             # Use all CPU cores\n",
    ")\n",
    "\n",
    "coords_2d = tsne_model.fit_transform(valid_embeddings)\n",
    "\n",
    "# Scale coordinates to fill the frame nicely (optional but helps visualization)\n",
    "# This creates that \"filled frame\" look\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler(feature_range=(-100, 100))\n",
    "coords_2d = scaler.fit_transform(coords_2d)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add coordinates as columns to df_clean\n",
    "df_clean['x_coords_tsne'] = coords_2d[:, 0]\n",
    "df_clean['y_coords_tsne'] = coords_2d[:, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-38.240646 , -27.92819  ],\n",
       "       [-54.59228  ,  76.619385 ],\n",
       "       [ 15.818377 , -33.132805 ],\n",
       "       ...,\n",
       "       [ 33.482437 ,  24.06134  ],\n",
       "       [ 62.346188 ,   2.6459503],\n",
       "       [ 63.095627 , -54.317375 ]], shape=(4199, 2), dtype=float32)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coords_2d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save coords. Coords cover only the non-orphans documents.\n",
    "df_clean[['uuid', 'UT', 'x_coords_tsne', 'y_coords_tsne']].to_csv(f'{tm_folder_path}/document_coords_tsne.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 591
    },
    "id": "I7E0_6bv9hBp",
    "outputId": "06faea2f-1c00-4ca7-958b-4dd2dcb10c36"
   },
   "outputs": [],
   "source": [
    "# Document information. Including the topic assignation\n",
    "dataset_clustering_results = topic_model.get_document_info(documents, df = corpus, metadata={\"Score\": scores})\n",
    "\n",
    "# Check for orphans (Topic == -1), save and remove them\n",
    "if -1 in dataset_clustering_results['Topic'].values:\n",
    "    orphans = dataset_clustering_results[dataset_clustering_results['Topic'] == -1]\n",
    "    orphans.to_csv(f'{tm_folder_path}/orphans.csv', index=False)\n",
    "    dataset_clustering_results = dataset_clustering_results[dataset_clustering_results['Topic'] != -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UT</th>\n",
       "      <th>uuid</th>\n",
       "      <th>Document</th>\n",
       "      <th>Topic</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>Representative_Docs</th>\n",
       "      <th>Top_n_words</th>\n",
       "      <th>Representative_document</th>\n",
       "      <th>Score</th>\n",
       "      <th>X_E</th>\n",
       "      <th>X_C</th>\n",
       "      <th>level0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>173</td>\n",
       "      <td>791aefbd-0fde-417e-8e1b-2f7210647032</td>\n",
       "      <td>Ethanol Whey not Cheese whey is classified as ...</td>\n",
       "      <td>27</td>\n",
       "      <td>27_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, production, in, lactose, mi...</td>\n",
       "      <td>[The Reform of Two Strains in Yogurt Since lot...</td>\n",
       "      <td>of - the - and - to - production - in - lactos...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.203861</td>\n",
       "      <td>0.203861</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>174</td>\n",
       "      <td>84d56c08-f62f-4bfd-986c-71a44c1af2ed</td>\n",
       "      <td>Bacman sequestering cadmium into Bacillus spor...</td>\n",
       "      <td>4</td>\n",
       "      <td>4_metal_heavy_heavy metal_metals</td>\n",
       "      <td>[metal, heavy, heavy metal, metals, the, and, ...</td>\n",
       "      <td>[Nanocrystal Ecoli Flocculation Union Heavy me...</td>\n",
       "      <td>metal - heavy - heavy metal - metals - the - a...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.490900</td>\n",
       "      <td>0.490900</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>175</td>\n",
       "      <td>8a69e820-ef66-4438-8ed2-08943ab2c6b1</td>\n",
       "      <td>Bacterial Relay Race In our project we aim at ...</td>\n",
       "      <td>25</td>\n",
       "      <td>25_the_of_to_in</td>\n",
       "      <td>[the, of, to, in, and, quorum, system, we, quo...</td>\n",
       "      <td>[A multilayer signalprocessing system based on...</td>\n",
       "      <td>the - of - to - in - and - quorum - system - w...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.079969</td>\n",
       "      <td>0.079969</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>176</td>\n",
       "      <td>572f8dd4-bb30-4d59-a319-2a5e650cce90</td>\n",
       "      <td>E coli Automatic Directed Evolution Machine a ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, for, in, evolution, synthet...</td>\n",
       "      <td>[rEvolver  Vivo La Evolution Naturallyoccurrin...</td>\n",
       "      <td>of - the - and - to - for - in - evolution - s...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.313584</td>\n",
       "      <td>0.313584</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>177</td>\n",
       "      <td>be21c9a1-0aac-4e87-86f5-f6023d374461</td>\n",
       "      <td>BacInVader  a new system for cancer genetic th...</td>\n",
       "      <td>10</td>\n",
       "      <td>10_cancer_tumor_cells_the</td>\n",
       "      <td>[cancer, tumor, cells, the, to, and, of, thera...</td>\n",
       "      <td>[BLAST  Bifidobacterium Longum induced Apoptos...</td>\n",
       "      <td>cancer - tumor - cells - the - to - and - of -...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.292083</td>\n",
       "      <td>0.292083</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UT                                  uuid  \\\n",
       "0  173  791aefbd-0fde-417e-8e1b-2f7210647032   \n",
       "1  174  84d56c08-f62f-4bfd-986c-71a44c1af2ed   \n",
       "2  175  8a69e820-ef66-4438-8ed2-08943ab2c6b1   \n",
       "3  176  572f8dd4-bb30-4d59-a319-2a5e650cce90   \n",
       "4  177  be21c9a1-0aac-4e87-86f5-f6023d374461   \n",
       "\n",
       "                                            Document  Topic  \\\n",
       "0  Ethanol Whey not Cheese whey is classified as ...     27   \n",
       "1  Bacman sequestering cadmium into Bacillus spor...      4   \n",
       "2  Bacterial Relay Race In our project we aim at ...     25   \n",
       "3  E coli Automatic Directed Evolution Machine a ...      6   \n",
       "4  BacInVader  a new system for cancer genetic th...     10   \n",
       "\n",
       "                               Name  \\\n",
       "0                  27_of_the_and_to   \n",
       "1  4_metal_heavy_heavy metal_metals   \n",
       "2                   25_the_of_to_in   \n",
       "3                   6_of_the_and_to   \n",
       "4         10_cancer_tumor_cells_the   \n",
       "\n",
       "                                      Representation  \\\n",
       "0  [of, the, and, to, production, in, lactose, mi...   \n",
       "1  [metal, heavy, heavy metal, metals, the, and, ...   \n",
       "2  [the, of, to, in, and, quorum, system, we, quo...   \n",
       "3  [of, the, and, to, for, in, evolution, synthet...   \n",
       "4  [cancer, tumor, cells, the, to, and, of, thera...   \n",
       "\n",
       "                                 Representative_Docs  \\\n",
       "0  [The Reform of Two Strains in Yogurt Since lot...   \n",
       "1  [Nanocrystal Ecoli Flocculation Union Heavy me...   \n",
       "2  [A multilayer signalprocessing system based on...   \n",
       "3  [rEvolver  Vivo La Evolution Naturallyoccurrin...   \n",
       "4  [BLAST  Bifidobacterium Longum induced Apoptos...   \n",
       "\n",
       "                                         Top_n_words  Representative_document  \\\n",
       "0  of - the - and - to - production - in - lactos...                    False   \n",
       "1  metal - heavy - heavy metal - metals - the - a...                    False   \n",
       "2  the - of - to - in - and - quorum - system - w...                    False   \n",
       "3  of - the - and - to - for - in - evolution - s...                    False   \n",
       "4  cancer - tumor - cells - the - to - and - of -...                    False   \n",
       "\n",
       "      Score       X_E  X_C  level0  \n",
       "0  0.203861  0.203861   28      28  \n",
       "1  0.490900  0.490900    5       5  \n",
       "2  0.079969  0.079969   26      26  \n",
       "3  0.313584  0.313584    7       7  \n",
       "4  0.292083  0.292083   11      11  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standar format for report analysis\n",
    "dataset_clustering_results = dataset_clustering_results.reset_index(drop=True)\n",
    "dataset_clustering_results = dataset_clustering_results.drop(columns=['text'])\n",
    "dataset_clustering_results['X_E'] = dataset_clustering_results['Score']\n",
    "dataset_clustering_results['X_C'] = dataset_clustering_results['Topic'] + 1\n",
    "\n",
    "# Assign 'level0' based on cluster coverage (90%)\n",
    "cluster_counts = dataset_clustering_results['X_C'].value_counts().sort_values(ascending=False)\n",
    "total_rows = len(dataset_clustering_results)\n",
    "cumulative = cluster_counts.cumsum() / total_rows\n",
    "main_clusters = cluster_counts.index[cumulative <= others_threshold].tolist()\n",
    "dataset_clustering_results['level0'] = dataset_clustering_results['X_C'].apply(lambda x: x if x in main_clusters else 99999)\n",
    "dataset_clustering_results.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UT</th>\n",
       "      <th>uuid</th>\n",
       "      <th>Document</th>\n",
       "      <th>Topic</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>Representative_Docs</th>\n",
       "      <th>Top_n_words</th>\n",
       "      <th>Representative_document</th>\n",
       "      <th>Score</th>\n",
       "      <th>X_E</th>\n",
       "      <th>X_C</th>\n",
       "      <th>level0</th>\n",
       "      <th>cl99</th>\n",
       "      <th>cl-99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>173</td>\n",
       "      <td>791aefbd-0fde-417e-8e1b-2f7210647032</td>\n",
       "      <td>Ethanol Whey not Cheese whey is classified as ...</td>\n",
       "      <td>27</td>\n",
       "      <td>27_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, production, in, lactose, mi...</td>\n",
       "      <td>[The Reform of Two Strains in Yogurt Since lot...</td>\n",
       "      <td>of - the - and - to - production - in - lactos...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.203861</td>\n",
       "      <td>0.203861</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>174</td>\n",
       "      <td>84d56c08-f62f-4bfd-986c-71a44c1af2ed</td>\n",
       "      <td>Bacman sequestering cadmium into Bacillus spor...</td>\n",
       "      <td>4</td>\n",
       "      <td>4_metal_heavy_heavy metal_metals</td>\n",
       "      <td>[metal, heavy, heavy metal, metals, the, and, ...</td>\n",
       "      <td>[Nanocrystal Ecoli Flocculation Union Heavy me...</td>\n",
       "      <td>metal - heavy - heavy metal - metals - the - a...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.490900</td>\n",
       "      <td>0.490900</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>175</td>\n",
       "      <td>8a69e820-ef66-4438-8ed2-08943ab2c6b1</td>\n",
       "      <td>Bacterial Relay Race In our project we aim at ...</td>\n",
       "      <td>25</td>\n",
       "      <td>25_the_of_to_in</td>\n",
       "      <td>[the, of, to, in, and, quorum, system, we, quo...</td>\n",
       "      <td>[A multilayer signalprocessing system based on...</td>\n",
       "      <td>the - of - to - in - and - quorum - system - w...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.079969</td>\n",
       "      <td>0.079969</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>176</td>\n",
       "      <td>572f8dd4-bb30-4d59-a319-2a5e650cce90</td>\n",
       "      <td>E coli Automatic Directed Evolution Machine a ...</td>\n",
       "      <td>6</td>\n",
       "      <td>6_of_the_and_to</td>\n",
       "      <td>[of, the, and, to, for, in, evolution, synthet...</td>\n",
       "      <td>[rEvolver  Vivo La Evolution Naturallyoccurrin...</td>\n",
       "      <td>of - the - and - to - for - in - evolution - s...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.313584</td>\n",
       "      <td>0.313584</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>177</td>\n",
       "      <td>be21c9a1-0aac-4e87-86f5-f6023d374461</td>\n",
       "      <td>BacInVader  a new system for cancer genetic th...</td>\n",
       "      <td>10</td>\n",
       "      <td>10_cancer_tumor_cells_the</td>\n",
       "      <td>[cancer, tumor, cells, the, to, and, of, thera...</td>\n",
       "      <td>[BLAST  Bifidobacterium Longum induced Apoptos...</td>\n",
       "      <td>cancer - tumor - cells - the - to - and - of -...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.292083</td>\n",
       "      <td>0.292083</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    UT                                  uuid  \\\n",
       "0  173  791aefbd-0fde-417e-8e1b-2f7210647032   \n",
       "1  174  84d56c08-f62f-4bfd-986c-71a44c1af2ed   \n",
       "2  175  8a69e820-ef66-4438-8ed2-08943ab2c6b1   \n",
       "3  176  572f8dd4-bb30-4d59-a319-2a5e650cce90   \n",
       "4  177  be21c9a1-0aac-4e87-86f5-f6023d374461   \n",
       "\n",
       "                                            Document  Topic  \\\n",
       "0  Ethanol Whey not Cheese whey is classified as ...     27   \n",
       "1  Bacman sequestering cadmium into Bacillus spor...      4   \n",
       "2  Bacterial Relay Race In our project we aim at ...     25   \n",
       "3  E coli Automatic Directed Evolution Machine a ...      6   \n",
       "4  BacInVader  a new system for cancer genetic th...     10   \n",
       "\n",
       "                               Name  \\\n",
       "0                  27_of_the_and_to   \n",
       "1  4_metal_heavy_heavy metal_metals   \n",
       "2                   25_the_of_to_in   \n",
       "3                   6_of_the_and_to   \n",
       "4         10_cancer_tumor_cells_the   \n",
       "\n",
       "                                      Representation  \\\n",
       "0  [of, the, and, to, production, in, lactose, mi...   \n",
       "1  [metal, heavy, heavy metal, metals, the, and, ...   \n",
       "2  [the, of, to, in, and, quorum, system, we, quo...   \n",
       "3  [of, the, and, to, for, in, evolution, synthet...   \n",
       "4  [cancer, tumor, cells, the, to, and, of, thera...   \n",
       "\n",
       "                                 Representative_Docs  \\\n",
       "0  [The Reform of Two Strains in Yogurt Since lot...   \n",
       "1  [Nanocrystal Ecoli Flocculation Union Heavy me...   \n",
       "2  [A multilayer signalprocessing system based on...   \n",
       "3  [rEvolver  Vivo La Evolution Naturallyoccurrin...   \n",
       "4  [BLAST  Bifidobacterium Longum induced Apoptos...   \n",
       "\n",
       "                                         Top_n_words  Representative_document  \\\n",
       "0  of - the - and - to - production - in - lactos...                    False   \n",
       "1  metal - heavy - heavy metal - metals - the - a...                    False   \n",
       "2  the - of - to - in - and - quorum - system - w...                    False   \n",
       "3  of - the - and - to - for - in - evolution - s...                    False   \n",
       "4  cancer - tumor - cells - the - to - and - of -...                    False   \n",
       "\n",
       "      Score       X_E  X_C  level0   cl99  cl-99  \n",
       "0  0.203861  0.203861   28      28  False  False  \n",
       "1  0.490900  0.490900    5       5  False  False  \n",
       "2  0.079969  0.079969   26      26  False  False  \n",
       "3  0.313584  0.313584    7       7  False  False  \n",
       "4  0.292083  0.292083   11      11  False  False  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standar format for report analysis\n",
    "dataset_clustering_results['cl99'] = [True if x == 99999 else False for x in dataset_clustering_results['level0']]\n",
    "dataset_clustering_results['cl-99'] = [True if x == 99999 else False for x in dataset_clustering_results['level0']]\n",
    "dataset_clustering_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "level0\n",
       "1     150\n",
       "2     122\n",
       "3     121\n",
       "4     120\n",
       "5     115\n",
       "     ... \n",
       "63     20\n",
       "64     19\n",
       "65     18\n",
       "66     18\n",
       "67     17\n",
       "Name: count, Length: 68, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_clustering_results.level0.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "id": "eyQDhrjyc6Oh"
   },
   "outputs": [],
   "source": [
    "# Save the dataframe\n",
    "dataset_clustering_results.to_csv(f'{tm_folder_path}/dataset_minimal.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "1RYcsXanZvrr",
    "outputId": "d625d86f-5897-4d75-8b32-54f2dd258ad3"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2026-01-19 16:37:24,304 - BERTopic - WARNING: When you use `pickle` to save/load a BERTopic model,please make sure that the environments in which you saveand load the model are **exactly** the same. The version of BERTopic,its dependencies, and python need to remain the same.\n"
     ]
    }
   ],
   "source": [
    "# Save the topic model\n",
    "topic_model.save(f'{tm_folder_path}/topic_model_object.pck')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       True\n",
       "1       True\n",
       "2       True\n",
       "3       True\n",
       "4       True\n",
       "        ... \n",
       "4194    True\n",
       "4195    True\n",
       "4196    True\n",
       "4197    True\n",
       "4198    True\n",
       "Name: uuid, Length: 4199, dtype: bool"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clean['uuid'].isin(dataset_clustering_results['uuid'])"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyM4oE6dpT6UwjIg42kzzsWo",
   "include_colab_link": true,
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
